---
title: "Using the Thornhill and Bayly Trees"
output: html_notebook
---

## Introduction

The following notebook uses Andrew Thronhill's (Thornhill et al. 2019) tree of ~675 Eucalypt taxa and Bayly et al.'s (2013) tree to make a combined supertree of as many taxa in our dataset as are included and then place the remaining trait-taxa to complete the tree we will use for any subsequent PGLS analysis and measurement of phylogenetic signal.

## Testing Currency of Taxon Names

```{r fig.height = 25, fig.width=20}
bayly.tree <- read.tree(file = "Input_data/Bayly_et_al_MPE_reSubmission_cp24_for_treebase_BL.nwk")
plot(bayly.tree)
plot(ggtree(bayly.tree) + geom_tiplab(size=4))
#to remove the duplicated tips
length(bayly.tree$tip.label)
bayly.tree <- keep.tip(bayly.tree, bayly.tree$tip.label[!grepl("2", bayly.tree$tip.label)])
plot(bayly.tree)
bayly.tree$tip.label <- gsub("_1", "", bayly.tree$tip.label)
bayly.tree$tip.label <- gsub("_", " ", bayly.tree$tip.label)
bayly.tips <- data.frame("taxon" = bayly.tree$tip.label)
#for namechecking
#write.csv(bayly.tree$tip.label, file = "bayly_names_for_check.csv")
```

All names current except E. deglupta, which similarly does not show up anywhere on APNI. However, it is on the Thornhill tree and considered an accepted name with Blume as the authority.

```{r fig.height= 120, fig.width=20}
thorn.tree <- read.nexus(file = "Input_data/Eucalypt_Bayes.tre")
# plot(thorn.tree)
# plot(ggtree(thorn.tree) + geom_tiplab(size=5))
#ggsave(ggtree(thorn.tree) + geom_tiplab(size=5), file = "thorntree.pdf", limitsize = FALSE, width = 25, height=100)
# thorn.tips <- data.frame("taxon" = thorn.tree$tip.label)
# grep("_[0-9]+", thorn.tips$taxon, value=TRUE)
# thorn.tips$taxon2 <- sub("_[0-9]+", "", thorn.tips$taxon)
# grep("_[A-Z].+$", thorn.tips$taxon2, value=TRUE)
# thorn.tips$taxon3 <- sub("_[A-Z].+$", "", thorn.tips$taxon2)
# grep("_[0-9]+$|_[A-Z].+$", thorn.tips$taxon, value=TRUE) #matches all those that have any capital letters or numbers at the end
# thorn.tips$taxon[!grepl("_[0-9]+$|_[A-Z].+$", thorn.tips$taxon)]
# thorn.tips$taxon4 <- sub("_[0-9]+$|_[A-Z].+$", "", thorn.tips$taxon) #gets rid of all the stuff on the end that isn't part of the taxon name
# test <- c("GB0127_335", "H8215B72")
# test1 <- sub("[0-9]+", "", test)
# test2 <- gsub("[0-9]+", "", test)
# test
# test1
# test2 # so sub only affects the first match within each element in the list but gsub affects every match within the entries regardless of how many there are

thorntips.for.nomcheck <- thorn.tree$tip.label
thorntips.for.nomcheck <- sub("_[0-9]+$|_[A-Z].+$", "", thorntips.for.nomcheck)
thorntips.for.nomcheck <- sub("subsp", "subsp.", thorntips.for.nomcheck)
thorntips.for.nomcheck <- sub("var", "var.", thorntips.for.nomcheck)
thorntips.for.nomcheck <- gsub("_", " ", thorntips.for.nomcheck)
#!!!CAREFUL NOT TO OVERWRITE WHAT THIS NEXT FILE AS IT CONTAINS THE SYNONYMS IN ANOTHER SHEET!!!
##write.csv(thorntips.for.nomcheck, file = "thorntips_for_check.csv", row.names = FALSE)
```

Tips that are already listed as synonyms:
Acmena smithii -> Syzigium smithii
Corymbia catenaria -> Corymbia watsoniana subsp. capillata
C. dampieri -> C. greeniana
C. dimorpha -> C. peltata
C. dolichocarpa -> C. clarksoniana
C. opaca -> C. terminalis
C. porphyritica -> C. ellipsoidea
C. variegata -> C. citriodora
E. corticosa -> E. argophloia
E. cuspidata -> E. angulosa
E. largiflorens -> E. bicolor
E. leptophylla -> E. foecunda
E. nitida -> E. ambigua
E. persistens -> E. tardecidens
E. petiolaris -> E. leucoxylon subsp. petiolaris
E. pilligaensis -> E. woollsiana
E. plauta -> E. dissimulata subsp. plauta
E. pleurocarpa -> E. tetragona
E. sparsifolia -> E. oblonga
E. yangoura -> E. globoidea
Hypocalymma linifolium -> Hypocalymma tetrapterum

After checking the status of both the current names of these synonyms above (according to the supplementary material in the thorhill paper) as well as the older names, there are many cases where both the older and newer names are current on APC or some of the names are not found. None of the relevant names correspind to tips of the Bayly tree, though some are in the trait dataset.

  Only E.cuspidata -> angulosa and E. yangoura -> E. globoidea need changing from the older names already being used in the thorn tips. However, both of the current names of these two taxa are already elsewhere in the tree so if we simply keep these tips as well, we can see where they are and determine if we want to keep them later.
  
  Otherwise the other taxa expected to showup on the namecheck as non-current based on not being those for which both older and newer names are current, after the two above have been changed:
    - C. catenaria
    - C. dampieri
    - C. dimorpha
    - C. dolichocarpa
    - C. porphyritica (strangely not actually in the thorn tips list, though the rest are)
    - C. variegata
    - E. pilligaensis
    - E. plauta (not found)
    - E. yangoura 

```{r}
thorn.tree$tip.label <- sub("_[0-9]+$|_[A-Z].+$", "", thorntips.for.nomcheck)
thorn.tree$tip.label <- sub("subsp", "subsp.", thorntips.for.nomcheck)
thorn.tree$tip.label <- sub("var", "var.", thorntips.for.nomcheck)
thorn.tree$tip.label <- gsub("_", " ", thorntips.for.nomcheck)
thorn.tips <- data.frame("taxon" = thorn.tree$tip.label)
# thorn.tree$tip.label[which(thorn.tree$tip.label == "Eucalyptus cuspidata")] <- "Eucalyptus angulosa"
# thorn.tree$tip.label[which(thorn.tree$tip.label == "Eucalyptus yangoura")] <- "Eucalyptus globoidea"
which(thorn.tree$tip.label == "Eucalyptus angulosa")
which(thorn.tree$tip.label == "Eucalyptus cuspidata") #would form paraphyletic group w.r.t. E. incrassata when extra taxa are pruned
#which(thorn.tree$tip.label == "Eucalyptus rugosa")
which(thorn.tree$tip.label == "Eucalyptus yangoura")
which(thorn.tree$tip.label == "Eucalyptus globoidea") # close but similarly not sister
#which(thorn.tree$tip.label == "Eucalyptus lateritica")
```

```{r}
#now to check the rest of the tips
thorntips.check <- as.data.frame(read.csv("Input_data/name_check_thorntips.csv", header = TRUE))
non.current.thorntips <- data.frame("taxon" = thorntips.check[!thorntips.check$Census == "APC", c("Search.term")])
sorted.synonyms.thorn <- c("Corymbia catenaria", "Corymbia dampieri", "Corymbia dimorpha", "Corymbia dolichocarpa", "Corymbia porphyritica", "Corymbia variegata", "Eucalyptus pilligaensis", "Eucalyptus plauta", "Eucalyptus yangoura")
non.current.thorntips2 <- data.frame("taxon" = non.current.thorntips$taxon[!non.current.thorntips$taxon %in% sorted.synonyms.thorn])
nc.thorntips.in.bayly <- data.frame("taxon"= non.current.thorntips2$taxon[non.current.thorntips2$taxon %in% bayly.tips]) #none in bayly bc all bayy tips are current
# non.current.thorntips$taxon[non.current.thorntips$taxon %in% sorted.synonyms.thorn]
#write.csv(non.current.thorntips2, file = "final_thorntips_for_check.csv", row.names = FALSE)

sum(all.taxa.clean$taxon %in% thorntips.for.nomcheck) #103 now
trait.taxa.nonthorn <- data.frame('taxon' = all.taxa.clean$taxon[!all.taxa.clean$taxon %in% thorntips.for.nomcheck]) #55 taxa not on the thornhill tree
sum(trait.taxa.nonthorn$taxon %in% bayly.tips$taxon) #hence of the 55 taxa not on the thorntree, only one of those is on the bayly tree instead (E. aromaphloia)

trait.taxa.nonthorn[trait.taxa.nonthorn$taxon %in% bayly.tips$taxon,] # E. aromaphloia is on the bayly tree instead
#will still need to add 54 taxa to these trees

non.matching.trait.taxa <- data.frame("taxon" = trait.taxa.nonthorn[!trait.taxa.nonthorn %in% bayly.tips$taxon,])

sum(!bayly.tips$taxon %in% thorntips.for.nomcheck) #11 so it's still worth checking the bayly tree as well
```

Thorn tips non-current and without synonym listed:

  Most of these are listed as 'not found' on APC so there's no way to know if they do correspond to more current names that are on the tree and so until there's more info on them, they can be left as pruned taxa and assumed to not have current names or synonyms matching trait taxa.
  
  For those that have a match on APC and therefore have observable taxonomic history:
  
  - E. caesia subsp. caesia -> E. caesia (in neither other tree)
  - E. argyphea -> E. falcata (already elsewhere on thorntree)
  - E. communalis (APC Excluded) -> no longer distinct and considered intergrade between E. adesmophloia and E. obesa(neither of which are in either of the other trees)
  - E. medialis -> E. hebetifolia (already elsewhere on thorntree)
  - E. brachycorys -> E. cometae-vallis (only on thorntree)
  - E. persistens subsp. tardecidens -> prob. E. tardecidens but ambiguous (could be E. persistens though unlikely) (neither current name is in either other tree)
  - E. apodophylla subsp. provecta -> E. apodophylla (in neither tree)
  - E. brunnea -> E. deanei (only on thorntree)
  - E. disclusa -> E. interstans (onlky on thorntree)
  - E. splendens subsp. splendens -> E. splendens (only on thorntree)
  - E. approximans subsp. codonocarpa -> E. codonocarpa (only on thorntree)
  - E. piperita subsp. piperita -> E. piperita (ON TRAIT DATASET!!)
  - E. andrewsii subsp. andrewsii -> E. andewsii (only on thorntree)
  - E. ralla -> E. tenella (in traits but elsewhere on thorntree)
  - E. oblonga -> globoidea or sparsifolia (!!!GLOBOIDEA in traits dataset and this taxon likely represents that bc there salready E. sparsifolia ni the thorntree)
  - E. eudesmioides subsp. eudesmioides -> E. eudesmioides (only on thorntree)
  - C. lignans subsp. lignans -> C. lignans (only on thorntree)
  - A. exul -> A. bakeri subsp. bakeri (already elsewhere on thorntree and in the trait dataset)
  
  - E. comitae vallis (misspelling of above E. cometae-vallis, hence why not found)

So from this above investigation, only E. piperita subsp. piperita needs to be changed to E. piperita on grounds of taxonomy and inclusion in the traits dataset.

```{r}
#taxonomic change relevant to trait taxon list
thorn.tree$tip.label[thorn.tree$tip.label == "Eucalyptus piperita subsp. piperita"] <- "Eucalyptus piperita"
```

```{r}
#just to check mis-spellings
not.found.thorntips <- data.frame("taxon" = thorntips.check[thorntips.check$Found. == "FALSE", c("Search.term")])
```

```{r}
#taxonomic change relevant to trait taxon list
thorn.tree$tip.label[thorn.tree$tip.label == "Eucalyptus decipiens subsp. decipens"] <- 'Eucalyptus decipiens subsp. decipiens'
```


## Adding Species and Species

Now to test which of the non-matching subspecies have species/other_sspp. that are simply resolved differently on the thorn tree or bayly tree.

```{r}
non.matching.trait.subspecies <- data.frame("taxon" = non.matching.trait.taxa[grepl("subsp.", non.matching.trait.taxa$taxon),])
nmt.spp.for.sspp <- data.frame("species" = str_split_fixed(non.matching.trait.subspecies$taxon, " ", 4)[,2])

thorn.tip.spp <- data.frame("species" = str_split_fixed(thorn.tips$taxon, " ", 4)[,2])

#now to check how many there are of the 28 mismatching sspp trait taxa have corresponding species in the thorn.tree
nmt.spp.for.sspp[nmt.spp.for.sspp$species %in% thorn.tip.spp$species,]
sum(nmt.spp.for.sspp$species %in% thorn.tip.spp$species)

sum(nmt.spp.for.sspp[nmt.spp.for.sspp$species %in% thorn.tip.spp$species,] == 'globulus') # 3
sum(nmt.spp.for.sspp[nmt.spp.for.sspp$species %in% thorn.tip.spp$species,] == 'radiata') # 1
sum(nmt.spp.for.sspp[nmt.spp.for.sspp$species %in% thorn.tip.spp$species,] == 'costata') # 1


#thorn.tips$taxon[thorn.tip.spp$species == "tricarpa"]
```

27/28 subsp. taxa from traits dataset have a species match on the thorn tree.
25 taxa with subspecies epithets non-mathcing from the triats dataset have same-species taxa on the thorntree (excl. costata, cinerea and decipiens):

5 pauciflora subspecies (non autonyms) on trait dataset have the autonymic subspecies as a representative on the thorn.tree, hence it's ok if they don't match since the autonymic subsp. is also in traits and is already goign to keep tip:
  - will need to add these manuall later

2 non-autonymic subsp. in traits represented by already set to match autonymic subsp. on the thorn.tree

E. calycogona subsp. trachybasis nearest represented by E. calycogona subsp. calycogona
  -must include this tip from thorn tree in pruned thorn tree

E. tricarpa subsp. tricarpa is species only on thorn tree
 
  .
  .
  .
      

```{r}
non.matching.trait.SPECIES <- data.frame("taxon" = non.matching.trait.taxa$taxon[!grepl("subsp.", non.matching.trait.taxa$taxon)])
#length(non.matching.trait.subspecies$taxon) + length(non.matching.trait.SPECIES$taxon) # ==55 so all good
nmt.spp.for.spp <- data.frame("species" = str_split_fixed(non.matching.trait.SPECIES$taxon, " ", 4)[,2])
nmt.spp.for.spp[nmt.spp.for.spp$species %in% thorn.tip.spp$species,]
sum(nmt.spp.for.spp$species %in% thorn.tip.spp$species)
```
10/27


## Matching Bayly and Thornhill Taxa

Need to figure out how many of the non-matching (with the thorntree) taxa in the Bayly tree have subspecies on the thorntree

```{r}
non.matching.bayly.tips <- data.frame("taxon" = bayly.tips$taxon[!bayly.tips$taxon %in% thorn.tips$taxon]) #all but 11 bayly tips have atches in the thorn tree

#now do those 11 have correspondent subspecies in the thorn tree
nmt.bayly.spp <- data.frame("species" = str_split_fixed(non.matching.bayly.tips$taxon, " ", 4)[,2])
nmt.bayly.spp[!nmt.bayly.spp$species %in% thorn.tip.spp$species,] 
sum(nmt.bayly.spp$species %in% thorn.tip.spp$species)
```

The two Bayly taxa that have no species match on the thorntree are the outgroups Allosyncarpia ternata and Stockwellia quadrifida.
The remaining 9 Bayly taxa with no match on the thorn tree, have a species epithet that does match thorn tree taxa.

```{r}
thorn.tips$taxon[thorn.tip.spp$species %in% nmt.bayly.spp$species] #finds thorn tips with species epithets matching bayly tips that don't otheriwise have a thorn tip match
```

All of those 9 taxa are indeed those with subspecies epithets and since, there are only 9, using only the species name shouldn't create too much more work later

Checking how many of theBayly tips don't match the trait taxa and how many of those are in the above list of 9 and will have to be changed back to their subspecies version that is actually in the trait dataset

How many of the 9 taxa have subspecies in the trait dataset?

```{r}
nmt.species <- rbind(nmt.spp.for.spp, nmt.spp.for.sspp)
nmt.species <- data.frame("species" = nmt.species$species[!duplicated(nmt.species$species)])

thorntip.sspp.for.nmt.bayly <- thorn.tips$taxon[thorn.tip.spp$species %in% nmt.bayly.spp$species]
thorntip.sspp.for.nmt.bayly.spp <- str_split_fixed(thorntip.sspp.for.nmt.bayly, " ", 4)[,2]

trait.taxa.subsp <- data.frame('taxon' = all.taxa.clean$taxon[grepl("subsp", all.taxa.clean$taxon)])
trait.taxa.spp.for.sspp <- str_split_fixed(trait.taxa.subsp$taxon, " ", 4)[,2]

trait.taxa.species <- data.frame('taxon' = all.taxa.clean$taxon[!grepl("subsp", all.taxa.clean$taxon)])
trait.taxa.spp.for.spp <- str_split_fixed(trait.taxa.species$taxon, " ", 4)[,2]

sum(thorntip.sspp.for.nmt.bayly.spp %in% trait.taxa.spp.for.sspp)
sum(trait.taxa.spp.for.sspp %in% thorntip.sspp.for.nmt.bayly.spp) #there are more this way around because the trait dataset can contain more than one subspecies with the same species epithet
trait.taxa.spp.for.sspp[trait.taxa.spp.for.sspp %in% thorntip.sspp.for.nmt.bayly.spp]
thorntip.sspp.for.nmt.bayly.spp[thorntip.sspp.for.nmt.bayly.spp %in% trait.taxa.spp.for.sspp]
thorntip.sspp.for.nmt.bayly.spp[!thorntip.sspp.for.nmt.bayly.spp %in% trait.taxa.spp.for.sspp]

# extra.3ofthe9 <- thorntip.sspp.for.nmt.bayly.spp[!thorntip.sspp.for.nmt.bayly.spp %in% trait.taxa.spp.for.sspp]
# 
# extra.3ofthe9[extra.3ofthe9 %in% trait.taxa.spp.for.spp] #aromaphloia so good
```

There are 6 taxa within these 9 that have subspecies on the traits dataset (of remaining 3, 2 aren't even in the traits dataset and the last is E. aromaphloia, which will match the species in the trait dataset anyway) that will need to be changed later.
Therefore there should be three taxa on this list of 9 bayly taxa that do match the trait dataset bc it also does not resolve them down to subspecies level

```{r}
thorn.tree$tip.label[thorn.tree$tip.label %in% thorntip.sspp.for.nmt.bayly] <- sub(" subsp.+$", "", thorn.tree$tip.label[thorn.tree$tip.label %in% thorntip.sspp.for.nmt.bayly])

# thorn.tips.change.test <- sub(" subsp.+$", "", thorn.tree$tip.label[thorn.tree$tip.label %in% thorntip.sspp.for.nmt.bayly])
# bayly.tips$taxon[bayly.tips$taxon %in% thorn.tips.change.test]


#now there should only be 2 non-matched between bayly tree and thorn tree
thorn.tips2 <- data.frame('taxon' = thorn.tree$tip.label)
bayly.tips$taxon[!bayly.tips$taxon %in% thorn.tips2$taxon] #good
```

##2nd Adding Species/Subspecies

Now to redo the namechecking with the traits dataset to see if the 18 taxa are still there:

```{r}
non.matching.trait.taxa2 <- data.frame('taxon' = all.taxa.clean$taxon[!all.taxa.clean$taxon %in% thorn.tips2$taxon]) # 60 (i.e. the 59 taxa from before (54 minus E. piperita) + the 6 taxa that now don't match because of the renaming so as expected)

#for the subspecies non-matching in the trait dataset
non.matching.trait.subspecies2 <- data.frame("taxon" = non.matching.trait.taxa2[grepl("subsp.", non.matching.trait.taxa2$taxon),])
nmt.spp.for.sspp2 <- data.frame("species" = str_split_fixed(non.matching.trait.subspecies2$taxon, " ", 4)[,2])

thorn.tip.spp2 <- data.frame("species" = str_split_fixed(thorn.tips2$taxon, " ", 4)[,2])

#now to check how many there are of the 34 (28 + 6) mismatching sspp trait taxa have corresponding species in the thorn.tree
nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] #all but syncarpia glomulifera
# sum(nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species)

nmt.spp.for.sspp2[!nmt.spp.for.sspp2$species %in% nmt.spp.for.sspp$species,] #camaldulensis, delegatensis, marginata
#only three new ones bc the other three taxa (that make up the six that now don't match already had subspecies that didnt match and so were already on the list as species and the above function doesn't return all duplicates

length(nmt.spp.for.sspp$species)
length(nmt.spp.for.sspp2$species)

nmt.spp.for.sspp2[!nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,]

sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "globulus") # now 4 not 3
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "radiata") # now 2 not 1
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "costata") # now 2 not 1

sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "pauciflora")      # 5
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "camaldulensis") 
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "viminalis")       # 2
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "calycogona") 
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "cinerea")         # 2
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "tricarpa")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "delegatensis")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "camphora")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "sideroxylon")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "alligatrix")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "baueriana")       # 2
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "leucoxylon")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "polyanthemos")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "tereticornis")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "marginata")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "ligulata")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "decipiens")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "conglobata")

nm.trait.sspp <- nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,]
nm.trait.sspp <- data.frame('species' = nm.trait.sspp[!duplicated(nm.trait.sspp)])
```

These 33/34 non-matching trait subspecies that have species epithets on the thorn tree (Syncarpia glomulifera subsp. glomulifera does not have any match at all on the thorn tree so will have to be dealth with some other way) can be positioned as follows:

  - Firstly we must rename/add those 6 names that were changed on the thorn tree to match the bayly tree:
      
      - E. camaldulensis -> E. camaldulensis subsp. camaldulensis (theres only one subspecies here)
      - E. globulus sspp. (there are 4) attach where E. globulus is attached
      - E. delegatensis -> E. delegatensis subsp. delegatensis
      - E. radiata sspp. (there are 2) atatch where E. radiata is positioned
      - E. marginata ->  E. marginata subsp. marginata (there's only one)
      - Angophora costata -> Angophora costata subsp. costata
 
  - For the rest:
  
      - E. pauciflora subspecies (there are 5) can be attached where E. pauciflora subsp. pauciflora is positioned
      - E. viminalis (there are 2) can be attached as sister to E. viminalis subsp. viminalis
      - E. calycogona subsp. trachybasis can be the new name of the autonymic thorn taxon subsp.
      - E. cinerea sspp. (there are 2) can be added and replace E. cinerea subsp. triplex 
      - E. tricarpa -> autonymic subsp.
      - E. camphora subsp. camphora -> E. camphora subsp. humeana
      - E. sideroxylon -> E. sideroxylon subsp. sideroxylon
      - E. alligatrix subsp. alligatrix -> E. alligatrix subsp. limaensis
      - E. baueriana sspp. (there are two) replace the E. baueriana tip
      - E. leucoxylon subsp. stephaniae can be added sister to E. leucoxylon subsp. leucoxylon from traits
      - E. polyanthemos subsp. polyanthemos -> E. polyanthemos subsp. vestita
      - E. tereticornis subsp. mediana can be added sister to E. tereticornis subsp. tereticornis from traits
      - E. marginata -> E. marginata subsp. marginata
      - E. ligulata subsp. ligulata -> E. ligulata subsp. stirlingica
      - E. decipiens subsp. decipiens replaced by E. decipiens + E. decipiens subsp. chalara
      - E. conglobata subsp. conglobata -> E. conglobata subsp. perata
      
  
Now for the non-subspecies trait taxa that don't match.

```{r}
non.matching.trait.SPECIES2 <- data.frame("taxon" = non.matching.trait.taxa2$taxon[!grepl("subsp.", non.matching.trait.taxa2$taxon)])
#length(non.matching.trait.subspecies2$taxon) + length(non.matching.trait.SPECIES2$taxon) # == 60 so all good
nmt.spp.for.spp2 <- data.frame("species" = str_split_fixed(non.matching.trait.SPECIES2$taxon, " ", 4)[,2])
nmt.spp.for.spp2[nmt.spp.for.spp2$species %in% thorn.tip.spp2$species,]
nmt.spp.for.spp2[!nmt.spp.for.spp2$species %in% thorn.tip.spp2$species,]

#nmt.spp.for.spp[nmt.spp.for.spp$species %in% thorn.tip.spp$species,]

sum(nmt.spp.for.spp2$species %in% thorn.tip.spp2$species) #8 so they all fit well and are two less than initially due to fixing piperita and Var. ovata
```




For these 9/26 trait species that have subspecies on the thorn tree:

  - The ovata problem should disappear once I change that name above - good it has disappeared
  - E. willisii subsp. willisii -> E. willisii
  - E. preissiana subsp. preissiana -> E. preissiana
  - E. wandoo subsp. wandoo -> E. wandoo
  - (E. decipiens subsp. decipiens -> E. decipiens if not already replaced by above substitution)
  - E. xanthonema subsp. xanthonema -> E. xanthonema
  - E. uncinata subsp. uncinata -> E. uncinata
  - E. flocktoniae subsp. flocktoniae -> E. flocktoniae
  - E. pluricaulis subsp. pluricaulis -> E. pluricaulis


Also, don't forget to: E. vegrandis subsp. vegrandis -> E. vergrandis in order to match with the trait dataset

To check for any species with more than one subspecies in rder to be sure of the monophyly of the species taxa:

```{r}
sum(duplicated(thorn.tips2)) #none so we know that no species has more than one representative on this tree
```

Hence we don't have to worry about some species having more multiple representatives in different places in the tree and then be uncertain where a new subspecies taxon should be placed bc the species itself is non-monophyletic; and the assumption that any subspecies is a representative of its species can hold and we can replace any subspecies with another or the species alone. 


For those remaining:

```{r}
nmt.spp.for.spp2[!nmt.spp.for.spp2$species %in% thorn.tip.spp2$species,]
```

For the 17 (these 17 + Syncarpia from the subspecies list that didn't even match a species, then minus the E. vergrandis here bc I know that only creates a problem bc it is mis-spelt on the trait dataset itself) that have no match of even species level:

- E. falciformis: nomeclatural synonym of basionym E. willisii subsp. falciformis (already have willisii on tree)

- E. sabulosa: nomenclatural synonym of E. aromaphloia subsp. sabulosa (already have aromaphloia on tree)

- E. silvestris considered under E. odorata but also sometimes under E. microcarpa (latter alread on tree) (Nicolle also has some subspecies of E. wimmerensis as synonyms)

- E. wimmerensis poorly resolved from E. viridis and E. polybractea and possibly E. cajuptea (the former two are on the thornhill tree so will be kept as tips as well)

- E. bunyip: nothing on APC but Nicolle considers it to be intergrade between E. camphora subsp. humeana and E. strzeleckii (both are taxa that either partially match through different subspecies or have only series-mate inference respectfully; hence can't place this species until E. strzeleckii is placed)

- E. punctata: some basionyms related to tereticornis and ovata (both on tree) but Nicolle has little info beyond larger list of series-mates

- Tristaniopsis laurina: not sure

- Syncarpia glomulifera: not sure

- E. imitans: very little info on APC or Nicolle but in same series as E. globoidea and E. conglomerata (both on thorn tree) along with a very long list of others

- E. rossii: was once called E. racemosa subsp rossii (racemosa on tree bc in traits and thorn), also in same series as E. haemostoma so may wish to also include the latter as well on the pruned tree

- E. chapmaniana: little information but Nicolle says it's in Viminales subser Circulares wih rubida and dalrympleana and co.)

- E. behriana: almost no info but same series as largiflorens (in trait dataset; also populnea and persistens both on thorn tree) (intertexta, orgadophila, sparsa, )(we'll leave the thorn tree memebrs of that series for now since only largiflorens is on the trait dataset)

- E. strzeleckii: no info on APC or Nicolle synonyms but same series as camphora, ovata, yarraensis, cadens aggregata, etc (all but cadens on the thorn tree but first tree on the trait dataset so we'll leave it at that since that's enought to determine monophyly)

- E. carolaniae: no APC info but Nicolle condiers it an intergrade between E. cypellocarpa and E. goniocalyx subsp. goniocalyx (both on tree and trait dataset)

- E. arenicola: no info, same series as willisii, elata, dives, croajingolensis, falciformis, radiata and others but we'll onily keep the ones already on the tree bc they're in the trait dataset bc there's enough there to determine how close they are

- E. lehmannii: orthographic variance causing non-match, listed as Eucalyptus lehmanii subsp. lehmanii

- E. vergrandis is a spelling mistake and is represented by E. vegrandis subsp. vegrandis

- E. doratoxylon: no APC info (aside from an obsolete synonym E. microstoma), Nicolle has its only series mate as E. decurva, which is already in the trait dataset so we we may be able to use that tips to place this taxon
  - BUT LO! it is written as E. dorOtoxylon on the thorntree with the o not the official a-form

Check names manually to catch orthographic variants:

none on the bayly tree

on the thorn tree:
- lehmannii written as lehmanii on thorn tree (another variant to that listed on APC, lehmanni)
- doratoxylon written as dorotoxylon on thorn tree
- (of course there's the instance of E. vergrandis misspelt on the trait dataset spelt correctly on the thorn tree)


Hence, of the 18 above (including E. vergrandis):

- 3 taxa mismatches result from orthographic mistakes or variance

- 2 taxa are poorly resolved from other taxa with at least one of those very similar taxa on the thorn tree
- 2 taxa are considered intergrades between other taxa that are on the tree, though the former of these species is an intergrade between taxa that themselves will be placed via indirect inference which compound the error in their placement)
- 3 taxa have clear synonyms where they were once considered subspecies of other current taxa that are also on the tree, which gives us some info of what their nearest relatives may be
- 1 other taxon has multiple synonyms related to two current taxa on the tree

- 2 taxa ara the outgroups Tristaniopsis and Syncarpia, which will require some digging in non-eucalypt literature to determine where they should be (none of the three taxon lists have any outgroups in common so noting where they fall for the trees to be merged after they are merged is not likely to be a reliable method to see how non-eucalypt taxa end up)
- 5 remaining taxa have other members of their series (as defined by Nicolle 2019) on the trait dataset or the thorntree so can potentially use that



Of the three obsolete names in the trait dataset:
(Already part of subspecies changing) E. cinerea subsp. triplex (current) is in the thorn.tree so will need to be changed back to the obsolete name E. cinerea subsp. victoriensis

The new name E. incrassata is on the thorn.tree and so a sister taxon for E. costata subsp. murrayana will need to be added once the trees are done, since E. incrassata is also a separate taxon on the traits dataset.

(already accounted for by above process) For E. decipiens ssp. chalara, the ssp. decipiens is on the thorn.tree but there's no E. adesmophloia between which this is an intergrade so this taxon will likely be added as a sister taxon to the E. decipiens tip.

Also remember to also keep E. cuspidata and E. yangoura to see where they fall, though i roughly know where that is anyway, since they are closeish but not sister (to the taxa that are also their current names) and so will likely end up paraphyletic or polyphyletic upon pruning.

```{r}
#extra taxa that we'll keep when we prune down the thorn tree
extra.taxa.for.keeping <- data.frame('taxon' = c('Eucalyptus yangoura', 'Eucalyptus cuspidata', 'Eucalyptus calycogona subsp. calycogona', 'Eucalyptus cinerea subsp. triplex', 'Eucalyptus tricarpa', 'Eucalyptus camphora subsp. camphora', 'Eucalyptus sideroxylon', 'Eucalyptus alligatrix subsp. alligatrix', 'Eucalyptus baueriana', 'Eucalyptus polyanthemos subsp. polyanthemos', 'Eucalyptus marginata', 'Eucalyptus ligulata subsp. ligulata', 'Eucalyptus decipiens subsp. decipiens', 'Eucalyptus conglobata subsp. conglobata', 'Eucalyptus vegrandis subsp. vegrandis', 'Eucalyptus willisii subsp. willisii', 'Eucalyptus preissiana subsp. preissiana', 'Eucalyptus wandoo subsp. wandoo', 'Eucalyptus xanthonema subsp. xanthonema', 'Eucalyptus uncinata subsp. uncinata', 'Eucalyptus flocktoniae subsp. flocktoniae', 'Eucalyptus pluricaulis subsp. pluricaulis', 'Eucalyptus lehmanii subsp. lehmanii', 'Eucalyptus dorotoxylon', 'Eucalyptus viridis', 'Eucalyptus polybractea', 'Eucalyptus canaliculata', 'Eucalyptus grisea', 'Eucalyptus propinqua', 'Eucalyptus major', 'Eucalyptus populnea', 'Eucalyptus sparsa', 'Eucalyptus brookeriana subsp. brookeriana', 'Eucalyptus apiculata', 'Eucalyptus burgessiana', 'Eucalyptus quadrangulata', 'Eucalyptus microcorys', 'Corymbia intermedia'))
length(extra.taxa.for.keeping$taxon) #38
extra.taxa.for.keeping$taxon[!extra.taxa.for.keeping$taxon %in% thorn.tips2$taxon] #good, none now
extra.taxa.for.keeping$taxon[extra.taxa.for.keeping$taxon %in% all.taxa.clean$taxon] #ashould be none if using original 158 taxa for trait dataset otherwise the final 3 with E. propinqua as well should be in the trait dataset once the new data has been incorporated (so 4)
```

##Final Thornhill

```{r}
#first to attach the trait dataset and bayly tree taxa together and extract duplicates
taxa.for.keeping <- as.data.frame(rbind(all.taxa.clean, bayly.tips)) #197
taxa.for.keeping.nd <- data.frame('taxon' = taxa.for.keeping[!duplicated(taxa.for.keeping$taxon),]) #184
taxa.for.keeping$taxon[duplicated(taxa.for.keeping$taxon)] #13 doubled taxa exsting in both bayly tree and trait dataset
sum(all.taxa.clean$taxon %in% bayly.tips$taxon) #13 so good
all.taxa.for.keeping <- as.data.frame(rbind(taxa.for.keeping.nd, extra.taxa.for.keeping)) #222
sum(extra.taxa.for.keeping$taxon %in% bayly.tips$taxon) #3
sum(extra.taxa.for.keeping$taxon %in% all.taxa.clean$taxon) #none, good
#extra.taxa.for.keeping[extra.taxa.for.keeping$taxon %in% bayly.tips$taxon,]
sum(extra.taxa.for.keeping$taxon %in% thorn.tips2$taxon) #all, good
all.taxa.for.keeping.clean <- data.frame('taxon' = all.taxa.for.keeping[!duplicated(all.taxa.for.keeping$taxon),]) #219
all.taxa.for.keeping$taxon[duplicated(all.taxa.for.keeping$taxon)] #those 3 in both bayly and extra taxa
#extra.taxa.for.keeping$taxon[extra.taxa.for.keeping$taxon %in% bayly.tips$taxon]
sum(!all.taxa.for.keeping.clean$taxon %in% thorn.tips2$taxon) #full list has an extra 2 that don't match bc the two outgroups of the bayly tree don't match
sum(!all.taxa.clean$taxon %in% thorn.tips2$taxon) #has two lees bc missing the two mismatches from the bayly outgroups
atfk.test <- data.frame('taxon'= all.taxa.for.keeping.clean$taxon[!all.taxa.for.keeping.clean$taxon %in% thorn.tips2$taxon])
atc.test <- data.frame('taxon' = all.taxa.clean$taxon[!all.taxa.clean$taxon %in% thorn.tips2$taxon])
atfk.test$taxon[!atfk.test$taxon %in% atc.test$taxon]
length(all.taxa.for.keeping.clean$taxon)

sum(!all.taxa.for.keeping.clean$taxon %in% thorn.tree$tip.label) #61 so good and there are 158 that do match (some of which are just extra and more than 1 for some taxa so dont look at the numbers too much here)
sum(thorn.tree$tip.label %in% all.taxa.for.keeping.clean$taxon) #158 so good
```



Now to prune the Thornhill Tree:

```{r fig.height=25, fig.width=15}
thorn.tree.pruned <- keep.tip(thorn.tree, thorn.tree$tip.label[thorn.tree$tip.label %in% all.taxa.for.keeping.clean$taxon])
length(thorn.tree.pruned$tip.label) #158 good
thorn.tree.pruned$edge.length <- NULL
plot_dim1(thorn.tree.pruned)
plot_dim1(bayly.tree)
bayly.tree$edge.length <- NULL
#plot_dim1(bayly.tree)
ggsave(ggtree(bayly.tree) + xlim(0, 20) + geom_tiplab(size=10, offset = 0), file = "Output_figures/bayly_tree.pdf", width = 15, height=30)
ggsave(ggtree(thorn.tree.pruned) + xlim(0, 45) + geom_tiplab(size=7), file = "Output_figures/thorn_tree_pruned.pdf", width = 15, height=45)
```


## Merging Trees into Supertree

```{r fig.height=25, fig.width=15}
trees.for.merging <- c(thorn.tree.pruned, bayly.tree)

set.seed(1) #to make sure that the trees we get are reproducible
consensus.trees100 <- replicate_supertree2(trees.for.merging, 100)
trees100 <-lapply(consensus.trees100, write.tree)
con.tree3.100 <-consensus(consensus.trees100, p= 0.5, check.labels = TRUE)
con.tree.strict <- consensus(consensus.trees100, p=1, check.labels = TRUE)
# plot(ggtree(con.tree3.100) + xlim(0, 35) + geom_tiplab(size=5))
# plot(ggtree(con.tree.strict) + xlim(0, 35) + geom_tiplab(size=5))
ggsave(ggtree(con.tree3.100) + xlim(0, 45) + geom_tiplab(size=5), file = "Output_figures/majrule_tree3_100.pdf", width = 15, height=45)
ggsave(ggtree(con.tree.strict) + xlim(0, 40) + geom_tiplab(size=5), file = "Output_figures/strict_tree3_100.pdf", width = 15, height=45)

plot(ggtree(con.tree3.100) + xlim(0, 40) + geom_tiplab(size=5))
plot(ggtree(con.tree.strict) + xlim(0, 40) + geom_tiplab(size=5)) #there is some difference especially in the moncalypts and radiatae series between the majority rules ns strict consensus trees
```

### Renaming and Finalising Tree

Now to edit the strict consensus tree so that the near-matched names match the trait dataset and extra subspecies can be added:

```{r fig.height=30, fig.width=15}
#starting with the tips corresponding to near-matching subspecies taxa in the traits dataset
merged.tree <- con.tree.strict
merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus camaldulensis"] <- 'Eucalyptus camaldulensis subsp. camaldulensis'

globulus.clade <- read.tree(text = "((Eucalyptus_globulus_subsp._globulus, Eucalyptus_globulus_subsp._bicostata, Eucalyptus_globulus_subsp._pseudoglobulus, Eucalyptus_globulus_subsp._maidenii));")
merged.tree <- replace_tip(merged.tree, "Eucalyptus globulus", globulus.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus delegatensis"] <- 'Eucalyptus delegatensis subsp. delegatensis'

radiata.clade <- read.tree(text = "(Eucalyptus_radiata_subsp._radiata, Eucalyptus_radiata_subsp._robertsonii);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus radiata", radiata.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus marginata"] <- 'Eucalyptus marginata subsp. marginata'

merged.tree$tip.label[merged.tree$tip.label == "Angophora costata"] <- 'Angophora costata subsp. costata'

pauciflora.clade <- read.tree(text = "(Eucalyptus_pauciflora_subsp._pauciflora, Eucalyptus_pauciflora_subsp._acerina, Eucalyptus_pauciflora_subsp._parvifructa, Eucalyptus_pauciflora_subsp._debeuzevillei, Eucalyptus_pauciflora_subsp._niphophila, Eucalyptus_pauciflora_subsp._hedraia);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus pauciflora subsp. pauciflora", pauciflora.clade)

viminalis.clade <- read.tree(text = "(Eucalyptus_viminalis_subsp._cygnetensis, Eucalyptus_viminalis_subsp._pryoriana, Eucalyptus_viminalis_subsp._viminalis);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus viminalis subsp. viminalis", viminalis.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus calycogona subsp. calycogona"] <- 'Eucalyptus calycogona subsp. trachybasis'

cinerea.clade <- read.tree(text = "(Eucalyptus_cinerea_subsp._victoriensis, Eucalyptus_cinerea_subsp._cinerea);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus cinerea subsp. triplex", cinerea.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus tricarpa"] <- 'Eucalyptus tricarpa subsp. tricarpa'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus camphora subsp. camphora"] <- 'Eucalyptus camphora subsp. humeana'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus sideroxylon"] <- 'Eucalyptus sideroxylon subsp. sideroxylon'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus alligatrix subsp. alligatrix"] <- 'Eucalyptus alligatrix subsp. limaensis'

baueriana.clade <- read.tree(text = "(Eucalyptus_baueriana_subsp._baueriana, Eucalyptus_baueriana_subsp._thalassina);")
merged.tree <- replace_tip(merged.tree,"Eucalyptus baueriana", baueriana.clade)

leucoxylon.clade <- read.tree(text = "(Eucalyptus_leucoxylon_subsp._stephaniae, Eucalyptus_leucoxylon_subsp._leucoxylon);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus leucoxylon subsp. leucoxylon", leucoxylon.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus polyanthemos subsp. polyanthemos"] <- 'Eucalyptus polyanthemos subsp. vestita'

tereticornis.clade <- read.tree(text = "(Eucalyptus_tereticornis_subsp._mediana, Eucalyptus_tereticornis_subsp._tereticornis);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus tereticornis subsp. tereticornis", tereticornis.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus ligulata subsp. ligulata"] <- 'Eucalyptus ligulata subsp. stirlingica'

decipiens.clade <- read.tree(text = "(Eucalyptus_decipiens, Eucalyptus_decipiens_subsp._chalara);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus decipiens subsp. decipiens", decipiens.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus conglobata subsp. conglobata"] <- 'Eucalyptus conglobata subsp. perata'

plot(ggtree(merged.tree) + xlim(0, 40) + geom_tiplab(size=5))
```

```{r fig.height=25, fig.width=15}
#now for the near-matched species-only traits taxa with members of the same species on the thorntree
# merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus willisii subsp. willisii"] <- 'Eucalyptus willisii'
# merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus preissiana subsp. preissiana"] <- 'Eucalyptus preissiana'
# merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus wandoo subsp. wandoo"] <- 'Eucalyptus wandoo'
# merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus xanthonema subsp. xanthonema"] <- 'Eucalyptus xanthonema'
# merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus uncinata subsp. uncinata"] <- 'Eucalyptus uncinata'
# merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus flocktoniae subsp. flocktoniae"] <- 'Eucalyptus flocktoniae'
# merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus pluricaulis subsp. pluricaulis"] <- 'Eucalyptus pluricaulis'

#or could be more efficient and simply sub-out the last bit all at once
sspp.to.spp <- c("Eucalyptus willisii subsp. willisii", "Eucalyptus preissiana subsp. preissiana", "Eucalyptus wandoo subsp. wandoo", "Eucalyptus xanthonema subsp. xanthonema", "Eucalyptus uncinata subsp. uncinata", "Eucalyptus flocktoniae subsp. flocktoniae", "Eucalyptus pluricaulis subsp. pluricaulis")
sspp.to.spp[sspp.to.spp %in% merged.tree$tip.label] #good all recognised
merged.tree$tip.label[merged.tree$tip.label %in% sspp.to.spp] <- sub(" subsp..+$", "", merged.tree$tip.label[merged.tree$tip.label %in% sspp.to.spp])
merged.tree$tip.label <- gsub("_", " ", merged.tree$tip.label)
```

Now to handle those of the 18 mismatches that we have agreed solutions for:

```{r}
#orthographic variants or mistakes
merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus vegrandis subsp. vegrandis"] <- 'Eucalyptus vergrandis'
merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus lehmanii subsp. lehmanii"] <- 'Eucalyptus lehmannii'
merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus dorotoxylon"] <- 'Eucalyptus doratoxylon'
```


```{r}
#the rule-taxa with basionyms that were sister subspecies to other taxa on the tree
falciformis.clade <- read.tree(text = "(Eucalyptus_falciformis, Eucalyptus_willisii);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus willisii", falciformis.clade)

sabulosa.clade <- read.tree(text = "(Eucalyptus_aromaphloia, Eucalyptus_sabulosa);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus aromaphloia", sabulosa.clade)

rossii.clade <- read.tree(text = "(Eucalyptus_rossii, Eucalyptus_racemosa);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus racemosa", rossii.clade)

merged.tree$tip.label <- gsub("_", " ", merged.tree$tip.label)

all.taxa.clean$taxon[!all.taxa.clean$taxon %in% merged.tree$tip.label] #once E. costata subsp. murrayana is dealt with (below) then there are only the 15 taxa (10 plus the two outgroups + 3 additional taxa not yet placed at this stage)
```

See what euclid says about the other 10 taxa that aren't outgroups or newly added taxa:

E. wimmerensis (only synonym E. viridis subsp. wimmerensis listed as vicflora synonym, very much a Rule-taxon), 
E. silvestris (Rule-taxon and considered by VicFlora to be somewhat intermediate between E. microcarpa and E. wimmerensis, also considered to be a depauperate but distinguishable form of E. microcarpa),
and E. carolaniae (vicflora has it having a synonym of E. aff. cypellocarpa (Mornington Peninsula) so maybe its closer to E. cypellocarpa than it is to E. goniocalyx and an be positioned accordingly) are nowhere to be found on Euclid and E. bunyip (VicFlora has no synonyms listed, it is a Rule taxon, but E. strzeleckii and E. camphora are noted as taxa it can be distinguished from) is noted as a synonym but not considered distinct enough to be included in EUCLID.



E. punctata has closest relatives, some of which are on the thornhill tree:
    - E. canaliculata (tiny coastal range north of Sydney)
    - E. grisea (tiny range in central QLD)
    - E. propinqua (similar range though not quite as south-reaching)
    - E. major (medium range all contained within QLD; potentially subspecies of E. propinqua)

```{r}
which(thorn.tree$tip.label == "Eucalyptus canaliculata")
which(thorn.tree$tip.label == "Eucalyptus grisea")
which(thorn.tree$tip.label == "Eucalyptus propinqua")
which(thorn.tree$tip.label == "Eucalyptus major")
```
Hence, they are relatively close on the thornhill tree to begin with and so it is likely they may even become sister to each other after it is pruned


Where was E. punctata collected?

```{r}
#full.se[full.se$taxon == "Eucalyptus punctata", c("locality")]
```


Samples were collected at the very southern tip of this species range, just south of Sydney and not far inland.

E. imitans a close relative of E. baxteri, which is on the traits dataset and already in the tree!

E. chapmaniana does not have any close relatives mentioned beyond how to tell it apart from certain other species.

E. behriana has close relatives E. populnea and E. sparsa (eastern and central australia repsectively). Can potentially say our sample is more likely to be closer to E. populnea based on the range in which it was collected (which was in SE Vic, Long Forest)

```{r}
# which(thorn.tree$tip.label == "Eucalyptus populnea")
# which(thorn.tree$tip.label == "Eucalyptus sparsa")
# #yeah very much that near each other
# 
# full.se[full.se$taxon == 'Eucalyptus behriana', c('locality')]
```

E. strzeleckii closest to E. yarraensis, E. ovata and E. brookeriana (latter has its autonym ssp. on the thornhill tree)

E. arenicola was once considered part of E. willisii subsp. willisii so can probably be placed as such nested within it or, more conservatively, as a polytomy with E. falciformis as well.



Now to add the extra 2 (that EUCLID named the closest relative of) accordingly:

```{r}
imitans.clade <- read.tree(text = "(Eucalyptus_imitans, Eucalyptus_baxteri);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus baxteri", imitans.clade)

arenicola.clade <- read.tree(text = "(Eucalyptus_falciformis, Eucalyptus_willisii, Eucalyptus_arenicola);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus willisii", arenicola.clade)
merged.tree <- drop.tip(merged.tree, c("Eucalyptus falciformis"))
```


Now also to remember to add E. costata murrayana as a sister taxon to E. incrassata, since Nicolle treats them as a single taxon, though APC considers them to both be current names.

```{r}
murrayana.clade <- read.tree(text = "(Eucalyptus_incrassata, Eucalyptus_costata_subsp._murrayana);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus incrassata", murrayana.clade)
merged.tree$tip.label <- gsub("_", " ", merged.tree$tip.label)
```


Now that we know what to keep as potential placeholders for these other 8 non-outgroup taxa, we can prune away the unecessary bayly species that aren't in the trait dataset:

```{r fig.height=30, fig.width=15}
extra.taxa.for.keeping2 <- data.frame("taxon" = c("Eucalyptus canaliculata", "Eucalyptus grisea", "Eucalyptus propinqua", "Eucalyptus major", "Eucalyptus viridis", "Eucalyptus polybractea", "Eucalyptus populnea", "Eucalyptus sparsa", "Eucalyptus brookeriana subsp. brookeriana", "Stockwellia quadrifida", "Eucalyptus apiculata", "Eucalyptus burgessiana", "Corymbia intermedia", "Eucalyptus quadrangulata", "Eucalyptus microcorys"))
extra.taxa.for.keeping2[!extra.taxa.for.keeping2$taxon %in% merged.tree$tip.label,] #none so good

merged.tree$tip.label <- gsub("_", " ", merged.tree$tip.label)
taxa.for.keeping2 <- data.frame("taxon" = rbind(all.taxa.clean, extra.taxa.for.keeping2))
sum(duplicated(taxa.for.keeping2$taxon)) #good none

merged.pruned.tree <- keep.tip(merged.tree, merged.tree$tip.label[merged.tree$tip.label %in% taxa.for.keeping2$taxon])

merged.pruned.tree$tip.label[merged.pruned.tree$tip.label == "Stockwellia quadrifida"] <- "Syncarpia glomulifera subsp. glomulifera"
plot_dim1(merged.pruned.tree)
```



Also, according to Thornhill, tribes Leptospermeae and chameliauciae are closest to Eucalypteae followed by Syncarpiae. Tristaniopsis and Lophostemon are in their own different tribes so of these three aforementioned tribes, only the Syncarpiae will have a member that we have data for and can be an outgroup sister to the eucalypteae. Hence, that can help us resolve the basal node of the tree by placing non-syncarpiae taxa as outgroups to the clade: (Syncapia + (Angophora + Corymbia + Eucalyptus)). None of the outgroups on the Thornhill or bayly trees are members of any of these other tribes so they can't be used for any positioning.

```{r fig.height=30, fig.width=15}
#Now to organise the outgroups with Syncarpia as the nearest non-eucalypt taxon
#write.tree(merged.pruned.tree)
final.tree1 <- read.tree(text = "((Tristaniopsis_laurina,Syncarpia_glomulifera_subsp._glomulifera));")
merged.pruned.tree$root.edge <- 0
final.tree1 <- bind.tree(final.tree1, merged.pruned.tree, where = 'root')
final.tree1 <- drop.tip(final.tree1, c("Syncarpia_glomulifera_subsp._glomulifera"))
final.tree1$tip.label <- gsub("_", " ", final.tree1$tip.label)
plot_dim1(final.tree1)

#ggsave(plot_dim1(final.tree1), file = "final_tree1.pdf", width = 15, height = 45)
```

In terms of any additional taxa being added from additional data and measurements
-Most were already in the traits dataset
-A couple (including Eucalyptus quadrangulata and potentially E. propinqua as well, are new to the dataset but in the Thornhill tree)
-Only one of them was not on the Thornhill tree:
  - E. dendromorpha (4/5 closest relatives on the thornhill tree)
  
  
  
### Testing Placement Effects   

So, at this point there are E. punctata (either next to propinqua or next to the more nested one of the other group), E. behriana that we will placed in various combination to see if it changes the answers.
Should we also do that with a few others such as E. dendromorpha (4 possible positions that are all different), E. wimmerensis (2 possible positions), E. silvestris (near wimmerensis or E. microcarpa), E. carolaniae (either near cypellocarpa or near goniocalyx), E. bunyip (near camphora subsp. humeana or near strzeleckii), E. strzeleckii (near one of the peppermints; ovata, yarraensis, brookeriana)

Also note which of these, if any, have extreme values for traits:
E. behriana has rather low values for reproductive traits but rather high stem density so can skew phylo-signal prbably the most 
E. punctata mostly intermediately-valued
E. wimmerensis has very small leaves, but intermediate SLA and missing data
E. silvestris has somehwhat low values for most traits but not particularly extreme
E. carolaniae has quite large leaves, though intermediate everyting else
E. bunyip has quite small reproductive characteristics but larger in everything else

E. strzeleckii has biggish leaf-traits and very small reproductive ones
E. chapmaniana has some large leaves in there

Since, we're not sure if we yet have the data for E. dendromorpha just yet:
- behriana next to largiflorens vs . next to populnea (sparsa isn't that far from populnea anyway so I can probabaly leave that one)
- wimmerensis (either next to viridis vs next to polybractea)
- silvestris next to wimmerensis and microcarpa (whch is sister to polybractea anyway) or not at all
- punctata (either next to propinqua or next to the nested one; keep behriana variable but have the others either both included or no silvestris)

```{r fig.height=30, fig.width=15}
#for the first combination:
  #behriana-largiflorens and wimmerensis-polybractea
behriana.clade1 <- read.tree(text = "(Eucalyptus_behriana, Eucalyptus_largiflorens);")
pot.tree1 <- replace_tip(final.tree1, "Eucalyptus largiflorens", behriana.clade1)
wimm.clade1 <- read.tree(text = "(Eucalyptus_polybractea, Eucalyptus_wimmerensis);")
pot.tree1 <- replace_tip(pot.tree1, "Eucalyptus polybractea", wimm.clade1)
pot.tree1$tip.label <- gsub("_", " ", pot.tree1$tip.label)
plot_dim1(pot.tree1)

#for behriana-largiflorens and wimmerensis-viridis
pot.tree2 <- replace_tip(final.tree1, "Eucalyptus largiflorens", behriana.clade1)
wimm.clade2 <- read.tree(text = "(Eucalyptus_wimmerensis, Eucalyptus_viridis);")
pot.tree2 <- replace_tip(pot.tree2, "Eucalyptus viridis", wimm.clade2)
pot.tree2$tip.label <- gsub("_", " ", pot.tree2$tip.label)

#for behriana-populnea and wimm-poly
behriana.clade2 <- read.tree(text = "(Eucalyptus_populnea, Eucalyptus_behriana);")
pot.tree3 <- replace_tip(final.tree1, "Eucalyptus populnea", behriana.clade2)
pot.tree3 <- replace_tip(pot.tree3, "Eucalyptus polybractea", wimm.clade1)
pot.tree3$tip.label <- gsub("_", " ", pot.tree3$tip.label)

#for behriana-populnea and wimm-viridis
pot.tree4 <- replace_tip(final.tree1, "Eucalyptus populnea", behriana.clade2)
pot.tree4 <- replace_tip(pot.tree4, "Eucalyptus viridis", wimm.clade2)
pot.tree4$tip.label <- gsub("_", " ", pot.tree4$tip.label)

# plot_dim1(pot.tree1)
# plot_dim1(pot.tree2)
# plot_dim1(pot.tree3)
# plot_dim1(pot.tree4)
# plot_dim1_nl(pot.tree1)

#for silvestris added onto where wimmerensis is with behriana with largiflorens
silvestris.clade <- read.tree(text = "(Eucalyptus_silvestris, Eucalyptus_wimmerensis, Eucalyptus_microcarpa);")
pot.tree5 <- replace_clade(pot.tree1, 'Eucalyptus wimmerensis', 'Eucalyptus microcarpa', silvestris.clade)
pot.tree5$tip.label <- gsub("_", " ", pot.tree5$tip.label)
plot_dim1(pot.tree5)

#for silvestris and wimmerensis in the same place but behriana with populnea
pot.tree6 <- replace_clade(pot.tree3, "Eucalyptus wimmerensis", "Eucalyptus microcarpa", silvestris.clade)
pot.tree6$tip.label <- gsub("_", " ", pot.tree6$tip.label)
plot_dim1(pot.tree6)

#for adding in punctata next to propinqua but behriana-largiflorens
punctata.clade1 <- read.tree(text = "(Eucalyptus_propinqua, Eucalyptus_punctata);")
pot.tree7 <- replace_tip(pot.tree1, "Eucalyptus propinqua", punctata.clade1)
pot.tree7$tip.label <- gsub("_", " ", pot.tree7$tip.label)
plot_dim1(pot.tree7)

#for adding punctata next to canaliculata but behriana-largiflorens
punctata.clade2 <- read.tree(text = "(Eucalyptus_canaliculata, Eucalyptus_punctata);")
pot.tree8 <- replace_tip(pot.tree1, 'Eucalyptus canaliculata', punctata.clade2)
pot.tree8$tip.label <- gsub("_", " ", pot.tree8$tip.label)
plot_dim1(pot.tree8)

#for adding punctata-propinqua but behriana-populnea
pot.tree9 <- replace_tip(pot.tree3, 'Eucalyptus propinqua', punctata.clade1)
pot.tree9$tip.label <- gsub("_", " ", pot.tree9$tip.label)
plot_dim1(pot.tree9)
#for adding punctata-canaliculata but behriana-populnea
pot.tree10 <- replace_tip(pot.tree3, "Eucalyptus canaliculata", punctata.clade2)
pot.tree10$tip.label <- gsub("_", " ", pot.tree10$tip.label)
plot_dim1(pot.tree10)

#for tree without behriana (keep wimmerensis and punctata and silvestris)
pot.tree11 <- pot.tree5 <- replace_clade(pot.tree9, 'Eucalyptus wimmerensis', 'Eucalyptus microcarpa', silvestris.clade)
pot.tree11 <- drop.tip(pot.tree11, "Eucalyptus behriana")
plot_dim1(pot.tree11)

#for the entire null tree
pot.null <- final.tree1
plot_dim1(pot.null)

#for behriana next to sparsa
behriana.clade3 <- read.tree(text = "(Eucalyptus_behriana, Eucalyptus_sparsa);")
pot.tree15 <- replace_tip(pot.tree5, "Eucalyptus sparsa", behriana.clade3)
pot.tree15 <- drop.tip(pot.tree15, c("Eucalyptus behriana"))
pot.tree15$tip.label <- gsub("_", " ", pot.tree15$tip.label)
plot_dim1(pot.tree15)

#now to prune all trees down to only the trait taxa and not the extra taxa included for placing (the pgls joining data into comparative objects does this anyway before running the analysis but i'm doing it here anyway so I can better see the effects of repositioning and what the tree used in each analysis actually looks like)
pot.trees <- c(pot.tree1, pot.tree2, pot.tree3, pot.tree4, pot.tree5, pot.tree6, pot.tree7, pot.tree8, pot.tree9, pot.tree10, pot.tree11, pot.null)

drop_non_trait_taxa <- function(tree) {
  drop.tip(tree, tree$tip.label[!tree$tip.label %in% all.taxa.clean$taxon])
}
pot.tree1.pruned <- drop_non_trait_taxa(pot.tree1)

all.taxa.clean$taxon[!all.taxa.clean$taxon %in% pot.tree1.pruned$tip.label]

plot_dim1(drop_non_trait_taxa(pot.tree1))

pot.trees.pruned <- lapply(pot.trees, drop_non_trait_taxa)

lapply(pot.trees.pruned, is.rooted)

# compute_brlen_grafen <- function(phy) {
#   compute.brlen(phy, method = "Grafen", power = 1)
# }


#pot.trees.pruned <- lapply(pot.trees.pruned, compute_brlen_grafen)
# for (i in 1:12) {
#   pot.tree.pruned.i <- pot.trees.pruned[[i]]
#   
# }
```



Careful bc E. quadrangulata (and potentially E. quadrangulata) will be in the traits dataset once it is added and hence it will not need to be put into the extra.taxa.for.keeping part.

```{r}
full.se.new <- read.csv("Input_data/traits_southeast.csv")
taxa.se.new <- distinct(data.frame('taxon' = full.se.new$taxon))
new.se.taxa <- taxa.se.new$taxon[!taxa.se.new$taxon %in% all.taxa.clean$taxon]
not.thornhill.new <- new.se.taxa[!new.se.taxa %in% thorn.tips2$taxon]
not.thornhill.new[not.thornhill.new %in% bayly.tips$taxon] #ok they aren't in Bayly's tree either so that didnt minimise them
```

Hence, the new data includes 4 taxa that are not on the thornhill tree: 
E. dendromorpha - we've decided not to include this taxon, since only 2 individuals were ever measured and there were no fruits measured

E. robusta - not sure but can check how many measurements were made using this to see how complete the data will be

```{r}
E.robusta.measurements <- as.data.frame(full.se.new[full.se.new$taxon == "Eucalyptus robusta",])
sort(E.robusta.measurements$trait[!duplicated(E.robusta.measurements$trait)])
```


So 4 individual trees of E. robusta are measured but there are no reproductive traits measured so maybe it's ok to exclude it from the tree since there isn't many correlations that could be drawn from it. Ask to check and do try and see (if I have time before next week) if I can find it in the Jones et al. tree since it is a Symp. (sect. Latoangulatae)

E. pauciflora - we decided to exclude this taxon from the se dataset originally, since it was not resolved down to subspecies level and there were already ample measurements for 6 sspp of this including the autonymic one most likely to be equivalent to this name.

Lophostemon confertus - we already know where this should go as an outgroup that is outermost instead of Tristaniopsis laurina

Of the 8 taxa that are not in the original 158:

E. pauciflora is not being included for aforementioned reaosns
E. dendromorpha- will also be excluded
E. robusta - may be excluded, depending on whether or not I can find an easy way to place it on the tree

The remaining 5 will need max. height added to them manually if they don't match the EUCLID data frame as well as the fire-response category:

E. quadrangulata
Corymbia intermedia
Lophostemon confertus
E. microcorys
E. propinqua


So after determining that the different positions of even the most extreme-valued dubious taxa (those whose position had been narrowed down at least to 2-4 positions) didn't greatly affect the measures of phylogenetic signal, and deciding its ok to only use Patrick's tree to place taxa not already on the thornhill tree and otherwise ignore its mild conflict about other taxa that are on it already, and deciding that it's ok to exclude E.dendromorpha bc it only has two individuals measured and no reproductive trait data, we need to:

- Place wimmerensis, silvestris and behriana where Patrick's tree says they go
- Make the outgroup Lophostemon instead of Tristaniopsis
- add punctata next to propinqua
```{r fig.height = 25, fig.width = 15}
#best to make this tree from scratch so
behriana.clade <- read.tree(text = "(Eucalyptus_behriana, Eucalyptus_largiflorens);")
final.tree2 <- replace_tip(final.tree1, "Eucalyptus largiflorens", behriana.clade)

wimm.clade <- read.tree(text = "(Eucalyptus_polybractea, Eucalyptus_wimmerensis);")
final.tree2 <- replace_tip(final.tree2, "Eucalyptus polybractea", wimm.clade)

silvestris.clade <- read.tree(text = "(Eucalyptus_silvestris, Eucalyptus_wimmerensis, Eucalyptus_microcarpa);")
final.tree2 <- replace_clade(final.tree2, 'Eucalyptus_wimmerensis', 'Eucalyptus microcarpa', silvestris.clade)

punctata.clade <- read.tree(text = "(Eucalyptus_propinqua, Eucalyptus_punctata);")
final.tree2 <- replace_tip(final.tree2, "Eucalyptus propinqua", punctata.clade)

final.tree2$tip.label[final.tree2$tip.label == "Tristaniopsis laurina"] <- "Lophostemon confertus"
final.tree2$tip.label <- gsub("_", " ", final.tree2$tip.label)

plot_dim1(final.tree2)
#pruning step
length(final.tree2$tip.label) #167
final.tree2$tip.label[!final.tree2$tip.label %in% all.taxa.clean$taxon]
new.taxa <- data.frame("taxon" = c("Lophostemon confertus", "Eucalyptus microcorys", "Eucalyptus propinqua", "Eucalyptus quadrangulata", "Corymbia intermedia"))
sum(!new.taxa$taxon %in% final.tree2$tip.label) #good none are not on that tree so they're spelled correctly
new.taxa.full <- data.frame('taxon' = rbind(all.taxa.clean, new.taxa))
final.tree2 <- keep.tip(final.tree2, final.tree2$tip.label[final.tree2$tip.label %in% new.taxa.full$taxon])
plot_dim1(final.tree2)
length(final.tree2$tip.label) #158 (there are original taxa missing as well as extra taxa (plus replaced taxa) so this is a coincidence)
```



Then do some reading on the remaining (chapmaniana, strzeleckii; also carolaniae and bunyip to see where they would be otherwise we can exclude them)

Firstly, punctata looks way more like it should go next to canaliculata and grisea (and longirostrata) rather than propinqua or major according to Jones et al. so maybe it should be placed there instead.

Jones et al. put E. strzeleckii (paraphyletic w.r.t. E. cadens) next to E. yarraensis (monophyletic), but should check Rule's reference anyway

E. chapmaniana may be able to be put next to E. smithii (Jones et al.) but it is also very close to E. glaucescens, which is in a different place, I think the same argument about why we can go with Patrick's tree only for the taxa not already on the thornhill tree, to justify why we can sort of ignore that that these two sources we're using have major conflicts in other parts of the tree even when those conflicts are only a taxon or two removed from the site of interest.

E. carolaniae and E. bunyip not on the Jones et al. tree so must find Rule's reference in order to determine where to place them if not entirely leaving them out
 
  Can't seem to access the article that Pete linked: https://www.rbg.vic.gov.au/documents/Muelleria_7(4),_p497-505,_Rule,_new_Eucalyptus.pdf 
  nor find some other site that has that same article that isn't on the missing part of the RBG site.

Now the taxa we expect to be missing from the tree despite being in the trait dataset:
- Tristaniopsis laurina
- E. dendromorpha
- maybe E. robusta too
(E. pauciflora should have been taken out at the start)

```{r fig.height = 25, fig.width = 15}
#best to make this tree from scratch so
behriana.clade <- read.tree(text = "(Eucalyptus_behriana, Eucalyptus_largiflorens);")
final.tree3 <- replace_tip(final.tree1, "Eucalyptus largiflorens", behriana.clade)

wimm.clade <- read.tree(text = "(Eucalyptus_polybractea, Eucalyptus_wimmerensis);")
final.tree3 <- replace_tip(final.tree3, "Eucalyptus polybractea", wimm.clade)

silvestris.clade <- read.tree(text = "(Eucalyptus_silvestris, Eucalyptus_wimmerensis, Eucalyptus_microcarpa);")
final.tree3 <- replace_clade(final.tree3, 'Eucalyptus_wimmerensis', 'Eucalyptus microcarpa', silvestris.clade)

punctata.clade <- read.tree(text = "(Eucalyptus_canaliculata, Eucalyptus_punctata);")
final.tree3 <- replace_tip(final.tree3, "Eucalyptus canaliculata", punctata.clade)

final.tree3$tip.label[final.tree3$tip.label == "Tristaniopsis laurina"] <- "Lophostemon confertus"


strzeleckii.clade <- read.tree(text = "(Eucalyptus_strzeleckii, Eucalyptus_yarraensis);")
final.tree3 <- replace_tip(final.tree3, "Eucalyptus yarraensis", strzeleckii.clade)

chapmaniana.clade <- read.tree(text = "(Eucalyptus_chapmaniana, Eucalyptus_smithii);")
final.tree3 <- replace_tip(final.tree3, "Eucalyptus smithii", chapmaniana.clade)


final.tree3$tip.label <- gsub("_", " ", final.tree3$tip.label)

plot_dim1(final.tree3)
#pruning step
length(final.tree3$tip.label)  #169
new.taxa <- data.frame("taxon" = c("Lophostemon confertus", "Eucalyptus microcorys", "Eucalyptus propinqua", "Eucalyptus quadrangulata", "Corymbia intermedia"))
sum(!new.taxa$taxon %in% final.tree3$tip.label) #good none are not on that tree so they're spelled correctly
new.taxa.full <- data.frame('taxon' = rbind(all.taxa.clean, new.taxa))
final.tree3 <- keep.tip(final.tree3, final.tree3$tip.label[final.tree3$tip.label %in% new.taxa.full$taxon])
plot_dim1(final.tree3)
length(final.tree3$tip.label) #160 (there are original taxa missing as well as extra taxa (plus replaced taxa) so this is a coincidence)
final.tree <- final.tree3
```

```{r}
#to export
ggsave(plot_dim1(final.tree), file = 'Output_figures/final_tree.pdf', width = 15, height = 25)
write.tree(final.tree, file = "Input_data/final_tree.txt") #lets put it here for now
```


## Heatmap of All Data

Now to plot visualise this data as a heatmap, plotting 3 different heatmaps for each group of traits: max height and stem sapwood density; SLA, leaf area and leaf mass; and fruit wall width, fruit mass, seed mass. This is so it easier to interpret patterns and read the map because it creates spaces between blocks of 2-3 related traits. Also data was log scaled (base 10) so that comparisons between traits with very different variances could be made.

```{r fig7, fig.height = 25, fig.width = 12}
# for 
final.tree <- read.tree("Input_data/final_tree.txt")
final.tree$tip.label <- gsub("_", " ", final.tree$tip.label)

#preparing the data for 
heatmap.data.height <- data.frame("MH" = eucs.scaled.data$max_height_m, "RH" = eucs.scaled.data$relative_height_by_girth, row.names = eucs.scaled.data$taxon)

heatmap.data.stem <- data.frame("SD" = eucs.scaled.data$stem_density_g_per_ml, "RBT" = eucs.scaled.data$relative_bt_by_girth, row.names = eucs.scaled.data$taxon)

heatmap.data.leaf <- data.frame("SLA" = eucs.scaled.data$sla_mm2_per_mg, "LA" = eucs.scaled.data$leaf_area_cm2, "LM" = eucs.scaled.data$leaf_mass_g, row.names = eucs.scaled.data$taxon)

heatmap.data.fruits <- data.frame("FW" = eucs.scaled.data$fruit_wall_width_mm, "FM" = eucs.scaled.data$fruit_mass_mg, "SM" = eucs.scaled.data$seed_mass_mg, row.names = eucs.scaled.data$taxon)

map.tree <- ggtree(final.tree, branch.length='none', layout='rectangular') %<+% eucs.scaled.data + xlim(0,70) + geom_tiplab(size = 4, offset = 26) + geom_tippoint(aes(colour=Sprouting_Type), size=2.7)
plot(map.tree)

map.height <- gheatmap(map.tree, heatmap.data.height,
  offset = -0.5,
  width = 0.2,
  low = "green",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Value"
) + xlim(0,70)

#plot(map.height)

map.stem <- gheatmap(map.height, heatmap.data.stem,
  offset = 4.5,
  width = 0.2,
  low = "green",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Value"
) + xlim(0,70)

plot(map.stem)

map.stem.leaf <- gheatmap(map.stem, heatmap.data.leaf,
  offset = 9.5,
  width = 0.3,
  low = "green",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Value"
) + xlim(0,70)

plot(map.stem.leaf)

map.total <- gheatmap(map.stem.leaf, heatmap.data.fruits,
  offset = 17,
  width = 0.3,
  low = "green",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Value"
) + xlim(0,70)

plot(map.total)

ggsave(map.total, file = "Output_figures/heatmap.pdf", width = 20, height = 25)
```

