---
title: "Using the Thornhill and Bayly Trees"
output: html_notebook
---

## Introduction

The following notebook uses Andrew Thronhill's (Thornhill et al. 2019) tree of ~675 Eucalypt taxa and Bayly et al.'s (2013) tree to make a combined supertree of as many taxa in our dataset as are included on them. Remaining taxa for which we have trait data will be placed on this tree using other smaller phylogenies and taxonomic information to complete the tree for any subsequent PGLS analysis and measurement of phylogenetic signal.


## Required Packages and Data

```{r}
# # To install ggtree package
#remotes::install_github("YuLab-SMU/ggtree") #let's see if this works?

library(ape)
library(BiocManager) # for installing ggtree but doesn't need to be loaded after that or even installed to  begin with if an alternate route of installing of ggtree
library(ggtree) # for more aesthetic visualisations of trees
library(ggplot2)
library(stringr)
library(phangorn) # contains the superTree() function used to combine various phylogenies to create the tree used in this analysis
library(caper)
library(viridis)

all.taxa.clean <- read.csv('../Input_data/all_taxa_clean.csv')
all.taxa.clean <- all.taxa.clean[c('taxon')]

eucs.scaled.data <- read.csv("../Input_data/eucs_scaled_data.csv")
load("~/Dropbox/Eucs_phylo_traits/Input_data/New_data/combined_traits.rda")
```

### Required Functions

```{r}
# shortcut fot plotting trees (of the relevant size for this analysis) with the right dimensions to be easily inspected in the output within the this markdown document
plot_dim1 <- function(phy) {
  # plots with the correct size text and alignment for 158-225 taxa trees like will be used
  plot(ggtree(phy) +xlim(0, 35) + geom_tiplab(size=5)) 
}

#the following function allows objects that are themselves comprised of a list of several objects to be repeated and each repeat be a single element of the list instead of contributing each of their constituent elements separately
rep_grouped <- function(X, n) {
  x <- list() 
  for (i in 1:n) {
    x[[i]] <- X
  }
  return(x)
}

# uses the above altered list-of-lists repetition function above to: repeat the process of creating a supertree from the same set of input trees a given number of times by list-applying the superTree() function to a list of identical lists of the inputs trees. In this version of this function, the superTree funtion is being applied to different but identical elements of a list of length n
replicate_supertree <- function(trees, n){
  x <- rep_grouped(trees, n)
  x <- lapply(x, function(treelist) {
    x1 <- superTree(treelist, method = "MRP", rooted = TRUE)
    x1$edge.length <- NULL
    return(x1)
    })
  return(x)
}

# does the same as the above funciton but a tad more efficiently. Here, instead of applying each superTree() interation to a differnt element of a list of identical sets of input trees, this time the function is iteratively being repeated and its output being stored as the single element of 
replicate_supertree2 <- function(trees, n) {
  l <- list()
for (i in 1:n) {
  tree <- superTree(trees, method = "MRP", rooted = TRUE)
  tree$edge.length <- NULL
  l[[i]] <- tree
}
  return(l)
}


# Required for following tip and clade replacement functions. Finds the node from which another node branches using tree (object of class 'phylo') and child.node (node number)
parent_node <- function(tree, child.node) {
  all.nodes <- as.data.frame(tree$edge)
  names(all.nodes) <- c("start", "end")
  parent.node <- all.nodes[all.nodes$end == child.node, c("start")]
  return(parent.node)
}


# the following allows the replacement of a single tip in a phylogeny with a clade of at least two related taxa (replacing a tip with another single tip can be more simply done by renaming a tip, for which a function already exists)
replace_tip <- function(tree, tip, new_clade) {
  # finds the tip-number of tip-taxon to be replaced
  node <- which(tree$tip.label == tip) 
  # adds id tag to new tips so that, in case one of the new tips is the same the the one being replaced (such as often occurs when adding sister taxa), we do in fact prune only the old tip at the end
  new_clade$tip.label <- paste0(new_clade$tip.label, "__new") 
  # sets root edge of new clade to be added onto the tree so that it goes to the right place and all new clades have standard behaviour under this function
  new_clade$root.edge <- 0 
  # adds the new clade as a sister taxon to the tip-to-be-replaced, attaching it at the parent node
  tree <- bind.tree(tree, new_clade, where = parent_node(tree, node)) 
  # deletes old tip
  tree <- drop.tip(tree, tip) 
  # removes new-tip ID tag
  tree$tip.label <- sub("__new", "", tree$tip.label) 
  return(tree)
}


#for finding the most recent common ancestor of two taxa; necessary for the replace_clade() function
pairwise_mrca <- function(tree, basal.taxon, tipward.taxon) {
  mrca.tree <- as.data.frame(mrca(tree))
  subset.mrca.tree <- data.frame(mrca.tree[row.names(mrca.tree) %in% c(basal.taxon, tipward.taxon), names(mrca.tree)[names(mrca.tree) %in% c(basal.taxon, tipward.taxon)]]) 
mrca.node <- min(subset.mrca.tree[subset.mrca.tree > length(tree$tip.label)])
return(mrca.node)
}

# the following allows the replacement of an entire clade with another clade of at least two related taxa, requires tree to be specified, two tips (as character) whose most recent common ancestor (MRCA) is the root of the clade to be replaced (to identify its node number) and the new clade (as phylo object) to be added as the replcement to the old clade
replace_clade <- function(tree, basal_clade_tip, tipward_clade_tip, new_clade) { 
  # finds node number of clade to be replaced by finding the msot recent common ancestor to two descendants, care to be taken to choose a pair of tips such that their MRCA is in fact the desired node
  clade_node <- pairwise_mrca(tree, basal_clade_tip, tipward_clade_tip)
  # storing the names of the current tips which we will prune later
  old_tips <- clade.members(clade_node, tree, tip.labels = TRUE)
  # adds ID tag to new tips so that, in case one of the new tips is the same the the one being replaced, we do in fact prune only the old tip at the end
  new_clade$tip.label <- paste0(new_clade$tip.label, "__new") 
  # sets root edge of new clade to be added onto the tree so that it goes to the right place and all new clades have standard behaviour under this function
  new_clade$root.edge <- 0 
  # adds the new clade as a sister taxon to the tip-to-be-replaced, attaching it at the parent node
  tree <- bind.tree(tree, new_clade, where = parent_node(tree, clade_node))
  # deletes old tips based on what the descendent tips were on the unaltered tree before the new clade was added i.e. any duplicate taxa on old an dnew clades can be differentiated using the '__new' tag
  tree <- drop.tip(tree, old_tips) 
  # removes new-tip ID tag
  tree$tip.label <- sub("__new", "", tree$tip.label) 
  return(tree)
}
```


## Testing Currency of Taxon Names

The following section seeks to check the names of taxa on both the Bayly and Thornhill trees in order to make sure taxa that differences in naming conventions don't cause taxa to be treated as seperate entities when they are in fact the same taxon. This was done by using the Australian Plant Names Index (APNI) and Australian Plant Census (APC) name-checking function (found here https://biodiversity.org.au/nsl/services/search/name-check) to determine where names on the Bayly and Thornhil trees were no longer current. Only taxa appearing on both trees and taxa with older synonyms appearing on the other tree that also have a synonym or current name in the trait dataset were looked at. This is because taxa not in the trait dataset or synonyms of taxa on the trait dataset will be pruned away eventually anyway as they are  not necessary for our analysis, and taxa that are only in one of the two phylogenies (assuming none of their synonyms are on the other tree) that will later be merged must only match the way that name is written in the trait dataset as there is no danger or creating diplicate tips for the same taxon upon merging the trees.

Since not every name has been manually checked after looking through its taxonomic history and determining the circumscription of the current name, there is a chance that names that are identical and current on different trees and the trait dataset may refer to a different subset of individuals.

```{r fig.height = 25, fig.width=20}
# for initial looking at the tree
bayly.tree <- read.tree(file = "../Input_data/Bayly_et_al_MPE_reSubmission_cp24_for_treebase_BL.nwk")
plot(bayly.tree)
plot(ggtree(bayly.tree) + geom_tiplab(size=4))

# to remove the duplicated tips (conveniently labelled with '_2' at the end)
length(bayly.tree$tip.label)
bayly.tree <- keep.tip(bayly.tree,
                  bayly.tree$tip.label[!grepl("2",bayly.tree$tip.label)])

# checking that it worked and those two tips are indeed missing
plot(bayly.tree)

# getting rid of numbers at the end of taxon names
bayly.tree$tip.label <- gsub("_1", "", bayly.tree$tip.label)
bayly.tree$tip.label <- gsub("_", " ", bayly.tree$tip.label)
bayly.tips <- data.frame("taxon" = bayly.tree$tip.label)

# for namechecking
# write.csv(bayly.tree$tip.label, file = "../Intermediate_Tree_Building_Products/bayly_names_for_check.csv")
```

All names current on the Bayly Tree except E. deglupta, which similarly does not show up anywhere on APNI. However, it is on the Thornhill tree and considered an accepted name with Blume as the authority.

```{r fig.height= 120, fig.width=20}
# reading in Thornhill's tree
thorn.tree <- read.nexus(file = "../Input_data/Eucalypt_Bayes.tre")

# getting rid of collecting IDs and other extra parts of names in order to create a list that can be put into APNI name-checking website
thorntips.for.nomcheck <- thorn.tree$tip.label
thorntips.for.nomcheck <- sub("_[0-9]+$|_[A-Z].+$", "", thorntips.for.nomcheck)
thorntips.for.nomcheck <- sub("subsp", "subsp.", thorntips.for.nomcheck)
thorntips.for.nomcheck <- sub("var", "var.", thorntips.for.nomcheck)
thorntips.for.nomcheck <- gsub("_", " ", thorntips.for.nomcheck)
#!!!CAREFUL NOT TO OVERWRITE THIS NEXT FILE, AS IT CONTAINS THE SYNONYMS IN ANOTHER SHEET!!!
# write.csv(thorntips.for.nomcheck, file = "../Intermediate_Tree_Building_Products/thornhill_taxa_for_APC.csv", row.names = FALSE)
# intermediate files such as name_check_thorntips.csv have the full output of the name-check on the full 675 in APC (in the Input_data folder); taxathorntips_for_check.csv and thorntips_for_check2.csv have the list of 20 taxa on the thorntree that were found using APC to have more current names
```

Many taxon names listed as 'not found' by the APC namecheck (most of the non-current ones are like this), some simply no longer current. Not all of the taxon names on the thornhill tree are (as currently written or their newer names) in the trait dataset so many will be pruned away later anyway.


Tips that are already listed as synonyms themselves:
Acmena smithii -> Syzigium smithii (this one isn't on the thorntree)
Corymbia catenaria -> Corymbia watsoniana subsp. capillata
C. dampieri -> C. greeniana
C. dimorpha -> C. peltata
C. dolichocarpa -> C. clarksoniana
C. opaca -> C. terminalis
C. porphyritica -> C. ellipsoidea
C. variegata -> C. citriodora
E. corticosa -> E. argophloia
E. cuspidata -> E. angulosa (latter in trait dataset)
E. largiflorens -> E. bicolor (former in trait dataset)
E. leptophylla -> E. foecunda (former in trait dataset, but there's already another tip with that former name)
E. nitida -> E. ambigua
E. persistens -> E. tardecidens
E. petiolaris -> E. leucoxylon subsp. petiolaris
E. pilligaensis -> E. woollsiana
E. plauta -> E. dissimulata subsp. plauta
E. pleurocarpa -> E. tetragona (former in trait dataset)
E. sparsifolia -> E. oblonga
E. yangoura -> E. globoidea (latter in trait dataset)
Hypocalymma linifolium -> Hypocalymma tetrapterum

After checking the status of both the current names of these synonyms above (according to the supplementary material in the thorhill paper) as well as the older names, there are many cases where both the older and newer names are current on APC or some of the names are not found. None of the relevant names correspond to tips of the Bayly tree, though some are in the trait dataset.

  Only E.cuspidata -> angulosa and E. yangoura -> E. globoidea need changing from the older names already being used in the thornhill tree tips. This is based on the fact that, of the 5 taxa that have either the current name or synonym on the trait dataset, only these two have the newer current name be the one that matches the trait dataset and so require the change so they match it. The other three taxa have the existing name match However, both of the current names of these two taxa are already elsewhere in the tree (there happens to be seperate tips on the tree for both taxon names) so if we simply keep these tips as well, we can see where they are and determine if we want to keep them later.
  
  Otherwise the other taxa expected to showup on the namecheck as non-current based on not being those for which both older and newer names are current (i.e. the ones where the latter name is supported as current by APC so we know what these are actually called now):
    - C. catenaria
    - C. dampieri
    - C. dimorpha
    - C. dolichocarpa
    - C. porphyritica (strangely not actually in the thorn tips list, though the rest are)
    - C. variegata
    - E. pilligaensis
    - E. plauta (not found)
    - E. yangoura 

```{r}
# re-formatting the taxon labels on the tree to something more readable. While spaces in names can be an issue in many places in r, it seems to not interfere too much with the functionality of the rest of the code
thorn.tree$tip.label <- sub("_[0-9]+$|_[A-Z].+$", "", thorntips.for.nomcheck)
thorn.tree$tip.label <- sub("subsp", "subsp.", thorntips.for.nomcheck)
thorn.tree$tip.label <- sub("var", "var.", thorntips.for.nomcheck)
thorn.tree$tip.label <- gsub("_", " ", thorntips.for.nomcheck)
thorn.tips <- data.frame("taxon" = thorn.tree$tip.label)

# Looking at where on the tree the current and superceded names are for the two taxa we will likely need to change the name of to see if they might be near or far from one another
which(thorn.tree$tip.label == "Eucalyptus angulosa")
## the following taxon would form paraphyletic group w.r.t. E. incrassata when extra taxa are pruned (sister on the tree)
which(thorn.tree$tip.label == "Eucalyptus cuspidata") 
## which(thorn.tree$tip.label == "Eucalyptus rugosa")
### the following two taxa are relatively close monocalypts
which(thorn.tree$tip.label == "Eucalyptus yangoura")
## close but similarly not sister
which(thorn.tree$tip.label == "Eucalyptus globoidea") 
# which(thorn.tree$tip.label == "Eucalyptus lateritica")
```

```{r}
# now to check the rest of the tips (excluding the ones where we already know what the current form of the name is meant to be)
thorntips.check <- as.data.frame(read.csv("../Input_data/name_check_thorntips.csv", header = TRUE))
non.current.thorntips <- data.frame("taxon" = thorntips.check[!thorntips.check$Census == "APC", c("Search.term")])
## now to remove the taxa we know the current names of and aren't a mystery we need to check. Note that Corymbia porphyritica is absent from the list to begin with so removing the following 9 taxa only reduces the taxon list by 8 not 9
sorted.synonyms.thorn <- c("Corymbia catenaria", "Corymbia dampieri", "Corymbia dimorpha", "Corymbia dolichocarpa", "Corymbia porphyritica", "Corymbia variegata", "Eucalyptus pilligaensis", "Eucalyptus plauta", "Eucalyptus yangoura")
non.current.thorntips2 <- data.frame("taxon" = non.current.thorntips$taxon[!non.current.thorntips$taxon %in% sorted.synonyms.thorn])

# Now that we know which versions of taxon names might be in the thornhill tree and the trait dataset, we need to check if any of the non-current taxon names in the thornhill tree are in the Bayly as we'll be merging them later
## none in bayly tree bc all bayly tips were current
non.current.thorntips$taxon[non.current.thorntips$taxon %in% sorted.synonyms.thorn]
non.current.thorntips2$taxon[non.current.thorntips2$taxon %in% bayly.tips] 
# write.csv(non.current.thorntips2, file = "final_thorntips_for_check.csv", row.names = FALSE)

# checking how many of our trait taxa match thornhill-tree taxa (thorntips.for.nomcheck contains all thornhill taxa before they were name-checked), 110 (was 107 before) taxa are in this category
sum(all.taxa.clean$taxon %in% thorntips.for.nomcheck) 
## the following indicates there are 56 (was 58 before) taxa in the trait dataset that are not on the thornhill tree at all
trait.taxa.nonthorn <- data.frame('taxon' = all.taxa.clean$taxon[!all.taxa.clean$taxon %in% thorn.tree$tip.label])
## the following indicates that, of the 58 taxa not on the thorntree, only one of those is on the bayly tree instead (E. aromaphloia)
sum(trait.taxa.nonthorn$taxon %in% bayly.tips$taxon) 
## now to see which species that single taxon on the bayly tree is, E. aromaphloia, will still need to add 53 (54-E. aromaphloia) taxa to these trees
trait.taxa.nonthorn[trait.taxa.nonthorn$taxon %in% bayly.tips$taxon,] 

non.matching.trait.taxa <- data.frame("taxon" = trait.taxa.nonthorn[!trait.taxa.nonthorn %in% bayly.tips$taxon,])

# checking if there are any tips on Bayly tree that are not on the thornhill tree, there's 11 so it's still worth checking the bayly tree as well. This is done in the next section.
sum(!bayly.tips$taxon %in% thorntips.for.nomcheck) 
```


Thorn tips non-current and without synonym listed:

  Most of these are listed as 'not found' on APC so there's no way to know if they do correspond to more current names that are on the tree and so until there's more info on them, they can be left as pruned taxa and assumed to not have current names or synonyms matching trait taxa.
  
  For those that have a match on APC and therefore have observable taxonomic history (and we can hence infer what the current name might be):
  
  - E. caesia subsp. caesia -> E. caesia (in neither other tree, so making sure the form of the name 'matches anything is unecessary)
  - E. argyphea -> E. falcata (already elsewhere on thorntree)
  - E. communalis (APC Excluded) -> no longer distinct and considered intergrade between E. adesmophloia and E. obesa (neither of which are in either of the other trees)
  - E. medialis -> E. hebetifolia (already elsewhere on thorntree)
  - E. brachycorys -> E. cometae-vallis (only on thorntree)
  - E. persistens subsp. tardecidens -> prob. E. tardecidens but ambiguous (could be E. persistens though unlikely) (neither current name is in either other tree)
  - E. apodophylla subsp. provecta -> E. apodophylla (in neither tree)
  - E. brunnea -> E. deanei (only on thorntree)
  - E. disclusa -> E. interstans (only on thorntree)
  - E. splendens subsp. splendens -> E. splendens (only on thorntree)
  - E. approximans subsp. codonocarpa -> E. codonocarpa (only on thorntree)
  - E. piperita subsp. piperita -> E. piperita (ON TRAIT DATASET!!)
  - E. andrewsii subsp. andrewsii -> E. andewsii (only on thorntree)
  - E. ralla -> E. tenella (in traits but elsewhere on thorntree)
  - E. oblonga -> globoidea or sparsifolia (E. globoidea in traits dataset and this taxon likely represents that bc thers already E. sparsifolia in the thorntree, but E. globoidea already exists as a seperate tip on the tree anyway so it doesn't require action)
  - E. eudesmioides subsp. eudesmioides -> E. eudesmioides (only on thorntree)
  - C. lignans subsp. lignans -> C. lignans (only on thorntree)
  - A. exul -> A. bakeri subsp. bakeri (already elsewhere on thorntree and in the trait dataset)
  
  - E. comitae vallis (misspelling of above E. cometae-vallis, hence why not found)

So from this above investigation, only E. piperita subsp. piperita needs to be changed to E. piperita on grounds of taxonomy and inclusion in the traits dataset. Especially considering we will be pruning down the thornhil tree before merging it with the bayly tree so taxa not included in that would not be relevant. Where there was already another tip with the current name, that existing tip can be used to represent the position of that trait-taxon.

```{r}
# taxonomic change relevant to trait taxon list
thorn.tree$tip.label[thorn.tree$tip.label == "Eucalyptus piperita subsp. piperita"] <- "Eucalyptus piperita"
```

```{r}
# just to check mis-spellings
not.found.thorntips <- data.frame("taxon" = thorntips.check[thorntips.check$Found. == "FALSE", c("Search.term")])
```
We will deal with these later, but the following on the above list of 28 'not found' taxa are relevant to our trait dataset:
E. dorotoxylon -> E. doratoxylon
E. decipiens subsp. decipens -> E. decipiens subsp. decipiens
E. lehmanii subsp. lehmanii -> E. lehmannii ...
E. ovata var ovata -> E. ovata subsp. ovata


```{r}
# taxonomic change relevant to trait taxon list based on miss-spelling on the thornhill tree causing this taxon to appear 'not found' (i.e. orthographic variants that will cause mis-matches with the trait dataset later)
thorn.tree$tip.label[thorn.tree$tip.label == "Eucalyptus decipiens subsp. decipens"] <- 'Eucalyptus decipiens subsp. decipiens'
thorn.tree$tip.label[thorn.tree$tip.label == "Eucalyptus lehmanii subsp. lehmanii"] <- 'Eucalyptus lehmannii subsp. lehmannii'
thorn.tree$tip.label[thorn.tree$tip.label == "Eucalyptus dorotoxylon"] <- 'Eucalyptus doratoxylon'
thorn.tree$tip.label[thorn.tree$tip.label == "Eucalyptus ovata var. ovata"] <- 'Eucalyptus ovata subsp. ovata'
```


## Matching Bayly and Thornhill Taxa

Need to figure out how many of the 11 non-matching (with the thorntree) taxa in the Bayly tree may not match simply because of different levels of resolution in the two trees. How many of those 11 taxa have member of the same species on the thorntree?

```{r}
# data frame of the 11 taxa in the Bayly tree that aren't found in the Thornhill tree
non.matching.bayly.tips <- data.frame("taxon" = bayly.tips$taxon[!bayly.tips$taxon %in% thorn.tips$taxon])

# now do those 11 have correspondent subspecies in the thorn tree
## first need to split the taxon names to isolate the species from the subspecies epithets
nmt.bayly.spp <- data.frame("species" = str_split_fixed(non.matching.bayly.tips$taxon, " ", 4)[,2]) 

# species epithets for all taxa on the thornhill tree
thorn.tip.spp <- data.frame("species" = str_split_fixed(thorn.tips$taxon, " ", 4)[,2])

# how many of the 11 taxa in nmt.bayly.spp have a match at the species level i.e. same species epithet, if not same subspecies epithet; 9 taxa of the 11 appear this way
sum(nmt.bayly.spp$species %in% thorn.tip.spp$species)

# Stockwellia quadrifida and Allosyncarpia ternata are the only ones that have no species-level match in the thornhill tree, which makes sense as these taxa are the outgroups that were used differently in the two papers/studies
nmt.bayly.spp[!nmt.bayly.spp$species %in% thorn.tip.spp$species,] 
```

The two Bayly taxa that have no species match on the thorntree are the outgroups Allosyncarpia ternata and Stockwellia quadrifida.
The remaining 9 Bayly taxa with no match on the thorn tree, but have a species epithet that does match thorn tree taxa.

```{r}
# finds how taxa with the same species but not subspecies epithets as bayly taxa are written on the thorntree
thorn.tips$taxon[thorn.tip.spp$species %in% nmt.bayly.spp$species] 
```

All of those 9 taxa are indeed those with subspecies epithets and since, there are only 9, using only the species name shouldn't create too much more work later to add the subspecies epithets back in

Now checking how many of the Bayly tips don't match the trait taxa and how many of those are in the above list of 9 and will have to be changed back to their subspecies version that is actually in the trait dataset

How many of the 9 taxa have subspecies in the trait dataset?

```{r}
# need to create lists of species names for taxa in the trait dataset that don't completely match tips in the thorntree, but for both the trait taxa that do and don't have 
## first to find the trait taxa NOT in the thornhill tree that are resolved to subspecies level
non.matching.trait.subspecies <- data.frame("taxon" = non.matching.trait.taxa[grepl("subsp.", non.matching.trait.taxa$taxon),])
## the following gives the species names for taxa in the trait dataset that have subspecies epithets and don't match the thornhill tree
nmt.spp.for.sspp <- data.frame("species" = str_split_fixed(non.matching.trait.subspecies$taxon, " ", 4)[,2])
## following line gives the species names for taxa in the trait dataset that do not have species epithets and don't match
nmt.spp.for.spp <- data.frame("species" = str_split_fixed(non.matching.trait.taxa$taxon[!grepl("subsp.", non.matching.trait.taxa$taxon)], " ", 4)[,2])
## now to make a total list of species names that exist in the trait dataset but not on the thorntree (there are 54 total as expected with 36 being trait subspecies taxa)
nmt.species <- rbind(nmt.spp.for.spp, nmt.spp.for.sspp)
## the following gets rid of the repeated species names from the trait dataset contaning multiple subspecies from the same species (reduces number down to 44 now)
nmt.species <- data.frame("species" = nmt.species$species[!duplicated(nmt.species$species)])

# creating vector of taxa on the thorntree corresponding to bayly taxa that don't match at the subspecies level
thorntip.spp.for.nmt.bayly <- thorn.tips$taxon[thorn.tip.spp$species %in% nmt.bayly.spp$species]
# creating vector of species epithets on the thorntree corresponding to bayly taxa that don't match at the subspecies level
thorntip.spp.for.nmt.bayly.spp <- str_split_fixed(thorntip.spp.for.nmt.bayly, " ", 4)[,2]

# created a column of all the trait taxa with subspecies epithets, 62 (used to be 50) taxa in total
trait.taxa.subsp <- data.frame('taxon' = all.taxa.clean$taxon[grepl("subsp", all.taxa.clean$taxon)])
# now extracting the species epithets for all trait taxa with subspecies epithets (regardless of matching to the thornhill tree), note this contains the duplicates
trait.taxa.spp.for.sspp <- str_split_fixed(trait.taxa.subsp$taxon, " ", 4)[,2]

# creating a column of the 102 (was 115) species in the trait dataset that DO NOT have subspecies epithets
trait.taxa.species <- data.frame('taxon' = all.taxa.clean$taxon[!grepl("subsp", all.taxa.clean$taxon)])
# creating a vector of species epithets for all trait taxa with no subspecies epithets
trait.taxa.spp.for.spp <- str_split_fixed(trait.taxa.species$taxon, " ", 4)[,2]

# checking how many of the 9 species on the thornhill tree (that match the unmatched bayly taxa at the species level but are differently resolved that Bayly taxa) are also in the list of subsp.-resolved non-matching trait taxa, there are 6 out of the 9
sum(thorntip.spp.for.nmt.bayly.spp %in% trait.taxa.spp.for.sspp)
# the following gives the species epithets of the 6 taxa
thorntip.spp.for.nmt.bayly.spp[thorntip.spp.for.nmt.bayly.spp %in% trait.taxa.spp.for.sspp]
# the following give the species epithets of the remaining 3 taxa that are not in the trait dataset 
thorntip.spp.for.nmt.bayly.spp[!thorntip.spp.for.nmt.bayly.spp %in% trait.taxa.spp.for.sspp]
```

There are 6 taxa within these 9 that have subspecies on the traits dataset (of remaining 3, 2 aren't even in the traits dataset and the last is E. aromaphloia, which will match the species in the trait dataset anyway) that will need to be changed back later to match the trait dataset.
Therefore, there should be three taxa on this list of 9 bayly taxa that do match the trait dataset bc it also does not resolve them down to subspecies level. We now convert all.

```{r}
# removing the subspecies epithets for the 9 thorntree taxa so they'll match the bayly tree taxa when those trees are merged
thorn.tree$tip.label[thorn.tree$tip.label %in% thorntip.spp.for.nmt.bayly] <- sub(" subsp.+$", "", thorn.tree$tip.label[thorn.tree$tip.label %in% thorntip.spp.for.nmt.bayly])


#now there should only be 2 non-matched between bayly tree and thorn tree
thorn.tips2 <- data.frame('taxon' = thorn.tree$tip.label)

# confirming they are the outgroups
bayly.tips$taxon[!bayly.tips$taxon %in% thorn.tips2$taxon] 
```

Bayly tips 'not' on Thornhill tree:

Allosyncarpia ternata -> not on thorntree but also not in trait dataset as it was simply a different outgroup
Stockwellia quadrifida -> not on thorntree but also not in trait dataset anyway so it will be pruned away regardless

The following 6/11 taxa are on the trait dataset and will have to have subspecies epithets added back in to match the trait dataset after the bayly and thornhill trees are merged:
Angophora costata -> on thorntree and traits as subsp. costata
E. camaldulensis -> on thorntree and traits as subsp. camaldulensis
E. delegatensis -> on Thorntree and traits as subsp. delegatensis
E. globulus -> on thorntree and traits as subsp. globulus
E. marginata -> on thorntree and traits as subsp. marginata
E. radiata -> on thorntree and traits as subsp. radiata

3 taxa in this list that are not in the trait dataset or are without a subspecies epithet:
E. aromaphloia -> on thorntree as subsp. aromaphloia, but on trait dataset as only the species epithet
E. diversifolia -> on thorntree as subsp. diversifolia but not in trait dataset
E. spathulata -> on thorntree as subsp. spathulata but not in trait dataset



## Adding Species and Subspecies

Many of the taxa in the trait dataset are resolved down to a different taxonomic level than the trees (i.e. subspecies when the tree only goes down to species level, or down to species level when the trees go down to subspecies). Yet, it appears that the Thornhill tree only includes a single member of each species even when there are multiple subspecies known, hence there is some assumption of species monophyly being made, especially in the cases where Thornhill tips are not resolved down to subspecies at all. As a result it is not particularly egregious for us to make the assumption that any member of a species can be used to represent the position of all members of that species i.e. species are monophyletic.
Now to test which of the non-matching subspecies have species/other_sspp. that are simply resolved differently on the thorn tree or bayly tree. This may not be the case for all species with no matches, but it is the case with the vast majority.

Now to redo the namechecking with the traits dataset to see if the 18 taxa are still there:

```{r}
# now to create a new list of trait taxa that don't match the thornhill tree names; 61 taxa now, 59 taxa we had before + the 6 taxa that have now been changed but minus the 4 thornhill taxa (E. piperita, E. doratoxylon, E. ovata + E. aromaphloia) that were changed to match the trait dataset
non.matching.trait.taxa2 <- data.frame('taxon' = all.taxa.clean$taxon[!all.taxa.clean$taxon %in% thorn.tips2$taxon]) 
# for the non-matching trait taxa, which have subspecies epithets?
non.matching.trait.subspecies2 <- data.frame("taxon" = non.matching.trait.taxa2[grepl("subsp.", non.matching.trait.taxa2$taxon),])
# the species epithets for such taxa
nmt.spp.for.sspp2 <- data.frame("species" = str_split_fixed(non.matching.trait.subspecies2$taxon, " ", 4)[,2])

# full thornhill list again but slightly altered
thorn.tip.spp2 <- data.frame("species" = str_split_fixed(thorn.tips2$taxon, " ", 4)[,2])

# now to check how many there are of the 41 (36 + 6 -1 for ovata being fixed) mismatching sspp trait taxa have corresponding species in the thorn.tree; 40 of 41 match (Syncarpia glomulifera the only one that doesn't as expected as thornhill used different outgroups to us)
nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,]
# to verify that it is the outgroup species Syncarpia glomulifera that does not match the thornhill tree
nmt.spp.for.sspp2[!nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,]


# camaldulensis, delegatensis, marginata returned by following call:
## only three new names bc the other three taxa (that make up the six that now don't match) already had subspecies that didnt match and so were already on the list as species and the above function only returns completely new names and not new duplicates of existing names
nmt.spp.for.sspp2[!nmt.spp.for.sspp2$species %in% nmt.spp.for.sspp$species,] 

# checking the number of subspecies for each unique species (28 unique species) within the 40/41 (i.e. excluding Syncarpia glomulifera) of nmt.spp.for.sspp2, the number is listed at the end of the line if it more than 1
nmt.spp.for.sspp2$species[duplicated(nmt.spp.for.sspp2$species)]
unique(nmt.spp.for.sspp2$species)

sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "globulus") # now 4 not 3
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "radiata") # now 2 not 1
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "costata") # now 2 not 1 (second from Angophora not Eucalyptus though, so not true subspecies in this case)

sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "pauciflora")      # 5
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "camaldulensis") 
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "viminalis")       # 2
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "calycogona") 
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "cinerea")         # 2
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "tricarpa")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "delegatensis")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "camphora")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "sideroxylon")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "alligatrix")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "baueriana")       # 2
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "leucoxylon") # 2 now from pruinosa also being added to the dataset
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "polyanthemos")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "tereticornis")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "marginata")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "ligulata")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "decipiens")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "conglobata")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "goniocalyx") # only 1 as the other subspecies was monotypic and so it matches with thornhill
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "lehmannii") # our only subspecies is not the one on thornhill
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "pluricaulis")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "rudis")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "thamnoides")
sum(nmt.spp.for.sspp2[nmt.spp.for.sspp2$species %in% thorn.tip.spp2$species,] == "xanthonema")
```

These 40/41 non-matching trait subspecies that have species epithets on the thorn tree (Syncarpia glomulifera subsp. glomulifera does not have any match at all on the thorn tree so will have to be dealt with and added to it in some other way) can be positioned as follows:

  - Firstly, once the two trees are merged (this will be done later) we will rename/add those 6 names back to what they are in the trait dataset in order to get them to match:
      
      - E. camaldulensis -> E. camaldulensis subsp. camaldulensis (only this subsp. in trait dataset)
      - E. globulus sspp. (there are 4 in the trait dataset) attach where E. globulus is attached
      - E. delegatensis -> E. delegatensis subsp. delegatensis (only this subsp. in trait dataset)
      - E. radiata sspp. (there are 2) atatch where E. radiata is positioned
      - E. marginata ->  E. marginata subsp. marginata (only this subsp. in trait dataset)
      - Angophora costata -> Angophora costata subsp. costata (only this subsp. in trait dataset)
 
 
  - For the rest:
  
      - E. pauciflora subspecies (there are 5) can be attached where E. pauciflora subsp. pauciflora is positioned (already matches a trait taxon)
      - E. viminalis (there are 2) can be attached as sister to E. viminalis subsp. viminalis
      - E. calycogona subsp. trachybasis can be the new name of the autonymic subsp. on thorntree (i.e. replacing the thornhill subsp. with the subsp. on the trait dataset we assume would have the same position relative to the rest of the taxa)
      - E. cinerea sspp. (there are 2) can be added and replace the position of E. cinerea subsp. triplex 
      - E. tricarpa -> autonymic subsp.
      - E. camphora subsp. camphora -> E. camphora subsp. humeana
      - E. sideroxylon -> E. sideroxylon subsp. sideroxylon
      - E. alligatrix subsp. alligatrix -> E. alligatrix subsp. limaensis
      - E. baueriana sspp. (there are two) replace the E. baueriana tip
      - E. leucoxylon subsp. stephaniae + E. leucoxylon subsp. pruinosa can be added to E. leucoxylon subsp. leucoxylon from trait dataset
      - E. polyanthemos subsp. polyanthemos -> E. polyanthemos subsp. vestita
      - E. tereticornis subsp. mediana can be added sister to E. tereticornis subsp. tereticornis from traits (don't need to keep this tip as it will already be kept when pruning the thornhill tree)
      - E. marginata -> E. marginata subsp. marginata
      - E. ligulata subsp. ligulata -> E. ligulata subsp. stirlingica
      - E. decipiens subsp. decipiens -> E. decipiens subsp. chalara
      - E. conglobata subsp. conglobata -> E. conglobata subsp. perata
      - E. goniocalyx subsp. viridissima can be added next to E. goniocalyx subsp. goniocalyx from the trait dataset
      - E. lehmannii subsp. lehmannii -> E. lehmannii subsp. parallela 
      - E. pluricaulis subsp. pluricaulis -> E. pluricaulis subsp. porphyrea
      - E. rudis -> E. rudis subsp. rudis
      - E. thamnoides -> E. thamnoides subsp. thamnoides
      - E. xanthonema subsp. xanthonema -> E. xanthonema subsp. apposita
  
Now for the non-subspecies trait taxa that don't match (i.e. trait taxa that thornhill resolve down to subspecies but the trait dataset does not):

```{r}
# data frame of the remaining 15 (18 originally -3 from E. aromaphloia, doratoxylon, and piperita) trait taxa (out of 56 that don't match the thornhill tree names) that do NOT have subspecies epithets
non.matching.trait.SPECIES <- data.frame("taxon" = non.matching.trait.taxa2$taxon[!grepl("subsp.", non.matching.trait.taxa2$taxon)])
## checking the length is 56 when combined to check code is doing what it should
length(non.matching.trait.subspecies2$taxon) + length(non.matching.trait.SPECIES$taxon) 

# now extracting their species epithets similarly to what was done for the other non-matching trait taxa, note there are no duplicates in this list of 28
nmt.spp.for.spp2 <- data.frame("species" = str_split_fixed(non.matching.trait.SPECIES$taxon, " ", 4)[,2])
# determining which of these species has a subspecies (another member of the species) on the thornhill tree, there are 3 and then the remaining 17 do not have species-level matches at all
nmt.spp.for.spp2[nmt.spp.for.spp2$species %in% thorn.tip.spp2$species,]
nmt.spp.for.spp2[!nmt.spp.for.spp2$species %in% thorn.tip.spp2$species,]
```


For these 3/20 trait species that have subspecies on the thorn tree, said subspecies can be renamed to get rid of the subspecies epithet in order to find the position of these species on the tree (i.e. they can use the position of the subspecies to decide where the species goes on he tree since we're assuming species monophyly):

  - E. willisii subsp. willisii -> E. willisii
  - E. uncinata subsp. uncinata -> E. uncinata
  - E. socialis subsp. socialis -> E. socialis


To check for any species with more than one subspecies in order to be sure of the monophyly of the species taxa:

```{r}
# none that aren't repeated species epithets in differnt genera (i.e. in Angophora or Corymbia) so we know that no species has more than one representative on this tree
sum(duplicated(thorn.tip.spp)) 
thorn.tips$taxon[duplicated(thorn.tip.spp)]
```

Hence we don't have to worry about some species having more multiple representatives in different places in the tree and then be uncertain where a new subspecies taxon should be placed bc the species itself is non-monophyletic; and the assumption that any subspecies is a representative of its species can hold and we can replace any subspecies with another or the species alone. 


For those remaining:

```{r}
nmt.spp.for.spp2[!nmt.spp.for.spp2$species %in% thorn.tip.spp2$species,]
```

For the 18 (these 17 (12 + 5 taxa we know will be subsequently removed) + Syncarpia from the subspecies list that didn't even match a species) that have no match of even species level:

- E. arenicola: no info, same series as willisii, elata, dives, croajingolensis, falciformis, radiata and others but we'll only keep the ones already on the tree bc they're in the trait dataset bc there's enough there to determine how close they are

- E. behriana: almost no info but same series as largiflorens (latter in trait dataset; also populnea and persistens both on thorn tree) (intertexta, orgadophila, sparsa)(we'll leave the thorn tree memebrs of that series for now since only largiflorens is on the trait dataset)

- E. chapmaniana: little information but Nicolle says it's in Viminales subser Circulares wih rubida and dalrympleana and co.)

- E. falciformis: nomenclatural synonym of basionym E. willisii subsp. falciformis (already have willisii on tree)

- E. imitans: very little info on APC or Nicolle but in same series as E. globoidea and E. conglomerata (both on thorn tree) along with a very long list of others

- E. punctata: some basionyms related to tereticornis and ovata (both on tree) but Nicolle has little info beyond larger list of series-mates

- E. rossii: was once called E. racemosa subsp rossii (racemosa on tree bc in traits and thorn), also in same series as E. haemostoma so may wish to also include the latter as well on the pruned tree

- E. sabulosa: nomenclatural synonym of E. aromaphloia subsp. sabulosa (already have aromaphloia on tree)

- E. silvestris considered under E. odorata but also sometimes under E. microcarpa (latter already on tree) (Nicolle also has some subspecies of E. wimmerensis as synonyms)

- E. strzeleckii: no info on APC or Nicolle synonyms but same series as camphora, ovata, yarraensis, cadens aggregata, etc (all but cadens on the thorn tree but first tree on the trait dataset so we'll leave it at that since that's enought to determine monophyly)

- E. wimmerensis poorly resolved from E. viridis and E. polybractea and possibly E. cajuptea (the former two are on the thornhill tree so will be kept as tips as well)

- Lophostemon confertus: not sure just yet. Need more information.

- Syncarpia lomulifer subsp. glomulifera: Not sure just yet. Need more information.

The list of `all.taxa.clean` used in this script already had the 4 taxa which we have subsequently decided to exclude based on there not being enough phylogenetic information and not enough actual trait data to justify investigating further into methods of attempting to place them. For completeness they are included in subsequent lists:

- E. bunyip: nothing on APC but Nicolle considers it to be intergrade between E. camphora subsp. humeana and E. strzeleckii (both are taxa that either partially match through different subspecies or have only series-mate inference respectfully; hence can't place this species until E. strzeleckii is placed)

- E. carolaniae: no APC info but Nicolle condiers it an intergrade between E. cypellocarpa and E. goniocalyx subsp. goniocalyx (both on tree and trait dataset)

- E. dendromorpha: unsure of position on the tree as of yet but some close relatives include E. apiculata and E. burgessiana

- E. robusta: unsure of position on the tree as of yet



Hence, of the 17 above + Syncarpia glomulifera:

- 3 taxa (falciformis, rossi, sabulosa) have clear synonyms where they were once considered subspecies of other current taxa that are also on the tree, which gives us some info of what their nearest relatives may be

- 2 taxa (silvestris and wimmerensis) are poorly resolved from other taxa with at least one of those very similar taxa on the thorn tree but can eventually decide where to put them

- 2 taxa (bunyip and carolaniae) are considered intergrades between other taxa often not sister to each other. Not even considered to be seperate entities by many official source and are hence difficult to position.

- 2 taxa (dendromorpha and robusta) have some relatives known but not much phylogenetic information.

- 1 other taxon (punctata) has multiple synonyms related to two current taxa on the tree

- 3 taxa ara the outgroups Lophostemon and Syncarpia, which will require some digging in non-eucalypt literature to determine where they should be (none of the three taxon lists have any outgroups in common so noting where they fall for the trees to be merged after they are merged is not likely to be a reliable method to see how non-eucalypt taxa end up)

- 5 remaining taxa (arenicola, behriana, chapmanniana, imitans, and strzeleckii) have other members of their series (as defined by Nicolle 2019) on the trait dataset or the thorntree so can potentially use that


Regarding each of the three obsolete names in the trait dataset:
(Already part of subspecies changing above so no further action required for this species) E. cinerea subsp. triplex (current) is in the thorn.tree so will need to be changed back to the obsolete name E. cinerea subsp. victoriensis to match the trait dataset again after trees have been merged (so that trait data will attach correctly)

The new name E. incrassata is on the thorn.tree and so a sister taxon for E. costata subsp. murrayana will need to be added once the trees are done, since E. incrassata is also a separate taxon on the traits dataset, despite these names being considered synonyms in other official sources.

(already accounted for by above process) For E. decipiens ssp. chalara, the ssp. decipiens is on the thorn.tree but there's no E. adesmophloia between which this is an intergrade so this taxon will be added as a sister taxon to the E. decipiens tip.



Also remember to keep E. cuspidata and E. yangoura to see where they fall, though I roughly know where that is anyway, since they are close-ish but not sister (to the taxa that are also their current names) and so will likely end up paraphyletic or polyphyletic upon pruning.

Given all of the above species names that will be used to position species based on other members of that same species, we need to determine which extra taxon names (i.e. tips) on the thorntree should be kept to anchor species without exact or intraspecific matches after we prune down the uneccessary majority of the taxa on the thornhill tree to reduce the computing speed. 

The following list also includes the possible nearest relatives of the remaining taxa that did not have members of the same species on the thornhill tree that could be used to position them.

Note: the six taxa on the thornhill tree that were originally resolved to subspecies level but were changed to species only to match the bayly tree don't need to be on the following list as the bayly tree taxa will be added to the list of thornhill taxa to keep anyway.

```{r}
# extra taxa that we'll keep when we prune down the thorn tree 
extra.taxa.for.keeping <- data.frame('taxon' = c('Eucalyptus yangoura', 'Eucalyptus cuspidata', 'Eucalyptus calycogona subsp. calycogona', 'Eucalyptus cinerea subsp. triplex', 'Eucalyptus tricarpa', 'Eucalyptus camphora subsp. camphora', 'Eucalyptus sideroxylon', 'Eucalyptus alligatrix subsp. alligatrix', 'Eucalyptus baueriana', 'Eucalyptus polyanthemos subsp. polyanthemos', 'Eucalyptus marginata', 'Eucalyptus ligulata subsp. ligulata', 'Eucalyptus decipiens subsp. decipiens', 'Eucalyptus conglobata subsp. conglobata', 'Eucalyptus lehmannii subsp. lehmannii', 'Eucalyptus pluricaulis subsp. pluricaulis', 'Eucalyptus rudis', 'Eucalyptus thamnoides', 'Eucalyptus xanthonema subsp. xanthonema', 'Eucalyptus willisii subsp. willisii', 'Eucalyptus uncinata subsp. uncinata', 'Eucalyptus socialis subsp. socialis', 'Eucalyptus polybractea', 'Eucalyptus canaliculata', 'Eucalyptus grisea', 'Eucalyptus major', 'Eucalyptus populnea', 'Eucalyptus sparsa', 'Eucalyptus brookeriana subsp. brookeriana', 'Eucalyptus apiculata', 'Eucalyptus burgessiana'))

# 31 names should be in this list
length(extra.taxa.for.keeping$taxon) 

# checking none of the above listed names are spelt wrong and do all match tips on the thornhill tree, should return 'character(0)'
extra.taxa.for.keeping$taxon[!extra.taxa.for.keeping$taxon %in% thorn.tips2$taxon] 

# making sure non of these names are redundant and allready in the trait dataset, should also return 'character(0)'
extra.taxa.for.keeping$taxon[extra.taxa.for.keeping$taxon %in% all.taxa.clean$taxon]
```

Other than the species with matches on the trait dataset at species level (i.e. partial intraspecific matches):

E. polybractea was kept as potential close relative of E. wimmerensis (other being E. viridis already on the tree)

E. canaliculata, E. grisea, and E. major were kept as potential close relatives of E. punctata ( in addition to E. propinqua which is already on the trait dataset)

E. populnea and E. sparsa were kept as potential close relatives of E. behriana.

E. brookeriana subsp. brookeriana was kept as as additional potential close relative of E. strzeleckii (in adition to E. yarraensis and E. ovata which allready match trait taxa)

E. apiculata and E. burgessiana were included as potential close relatives of E. dendromorpha (we eventually decide to exclude this taxon)

It is worth noting that there are several other taxa whose positioning will be done later if they are to be included at all: 
- Outgroups such as Syncarpia glomulifera subsp. glomulifera, Lophostemon confertus, Tristaniopsis laurina
- E. robusta, E. bunyip, E. carolaniae (taxonomically uncertain relationships)


## Final Thornhill

After all checking of names and mismatches causes by differences in taxonomic resolution, spelling variation, or the lack of direct represenatatives of trait taxa on the thornhill tree, we actually prune the thornhill tree down to the smallest st of taxa necessary for our analysis before merging it with the bayly tree.


```{r}
# first to attach the trait dataset and bayly tree taxa together and extract duplicates, it should contain 208 names including the duplicates, as we want to make sure none of the taxa on the bayly tree get excluded from the pruned thornhill tree even if they're not in the trait dataset
taxa.for.keeping <- as.data.frame(rbind(all.taxa.clean, bayly.tips))
## the follow column contains 193 names
taxa.for.keeping.nd <- data.frame('taxon' = taxa.for.keeping[!duplicated(taxa.for.keeping$taxon),]) 
## 15 duplicated taxa (existing in both bayly tree and trait dataset)
taxa.for.keeping$taxon[duplicated(taxa.for.keeping$taxon)] 
sum(all.taxa.clean$taxon %in% bayly.tips$taxon) 
# the following should be 224 names long
all.taxa.for.keeping <- as.data.frame(rbind(taxa.for.keeping.nd, extra.taxa.for.keeping))

# 2 taxa in the extra list were in the bayly tree, marginata and polybractea
sum(extra.taxa.for.keeping$taxon %in% bayly.tips$taxon) 
extra.taxa.for.keeping[extra.taxa.for.keeping$taxon %in% bayly.tips$taxon,]

# checking again for no redundancy or spelling mistakes, should show none and all 31 of them respectively
sum(extra.taxa.for.keeping$taxon %in% all.taxa.clean$taxon) #none, good
sum(extra.taxa.for.keeping$taxon %in% thorn.tips2$taxon) #all, good

# without duplicates this should be 222 names long, after we get rid of the 8 duplicated taxa
all.taxa.for.keeping.clean <- data.frame('taxon' = all.taxa.for.keeping[!duplicated(all.taxa.for.keeping$taxon),])
all.taxa.for.keeping$taxon[duplicated(all.taxa.for.keeping$taxon)] 
# bayly tips added an extra 2 outgroup taxa to this list of taxa that don't match any tips in the thornhill tree, the latter line should read 61 as it was originally and the first line should be two more long than the second line
sum(!all.taxa.for.keeping.clean$taxon %in% thorn.tips2$taxon)
sum(!all.taxa.clean$taxon %in% thorn.tips2$taxon) 

# testing which taxa in the above list of 63 are not in the thornhill tree (and subsequently confirming it is the outgroups of the bayly treee that account for the difference)
atfk.test <- data.frame('taxon'= all.taxa.for.keeping.clean$taxon[!all.taxa.for.keeping.clean$taxon %in% thorn.tips2$taxon])
# compared with the 61 taxa in the trait dataset that are not 
atc.test <- data.frame('taxon' = all.taxa.clean$taxon[!all.taxa.clean$taxon %in% thorn.tips2$taxon])
# shows that the two extra taxa that now also don't match the thornhill tree indeed come from the bayly tree outgroups
atfk.test$taxon[!atfk.test$taxon %in% atc.test$taxon]
atfk.test$taxon[!atfk.test$taxon %in% non.matching.trait.taxa2$taxon]

# there are 222 names currently in this list of trait taxa + the species we will use to position most of those that don't have a direct match
length(all.taxa.for.keeping.clean$taxon)

# 159 taxa of these 222 (222-63) match a tip on the thornhill tree the rest are trait taxa that we've included close relatives or other subspecies for to anchor them
sum(all.taxa.for.keeping.clean$taxon %in% thorn.tips2$taxon)
```



Now to prune the Thornhill Tree:

```{r fig.height=25, fig.width=15}
# only keep tips on thornhill tree that are in the above list
thorn.tree.pruned <- keep.tip(thorn.tree, thorn.tree$tip.label[thorn.tree$tip.label %in% all.taxa.for.keeping.clean$taxon])

# check the number of tips is in fact the 159 as we expect
length(thorn.tree.pruned$tip.label)

# removes branch lengths on the tree in preparation to combine it with the bayly tree that likely uses a different root age and possibly also units for branch length
thorn.tree.pruned$edge.length <- NULL
plot_dim1(thorn.tree.pruned)
ggsave(ggtree(thorn.tree.pruned) + xlim(0, 45) + geom_tiplab(size=7), file = "../Output_figures/thorn_tree_pruned.pdf", width = 15, height=45)

# removed branch length from the bayly tree as well
bayly.tree$edge.length <- NULL
plot_dim1(bayly.tree)
ggsave(ggtree(bayly.tree) + xlim(0, 20) + geom_tiplab(size=10, offset = 0), file = "../Output_figures/bayly_tree.pdf", width = 15, height=30)
```


## Merging Trees into Supertree

Here we use Supertree methods to merge the Bayly et al. (2013) tree with the (somewhat pruned) Thornhill et al. (2019) tree, which allows combination and 'consensus' between phylogenies of only partially overlapping taxa. Below, we use the Matrix Represenatation Parsimony method for choosing the best output tree.

```{r fig.height=25, fig.width=15}
# vector of the two tree objects to be merged
trees.for.merging <- c(thorn.tree.pruned, bayly.tree)

# to make sure that the trees we get are reproducible we need to set a seed for the random component of how the supertree function resolves relationships in the regions of uncertainty
set.seed(1) 
# here we are using one of 2 functions that merges the trees 100 times and stores each of the 100 outputs as single entries in a list
consensus.trees100 <- replicate_supertree2(trees.for.merging, 100)
trees100 <-lapply(consensus.trees100, write.tree)

# nice to have a look at what each of these trees look like and what the differences between them are, difficult to do in the Rmarkdown but a bit easier in pdf
pdf("../Output_figures/individual_trees_for_consensus.pdf", height = 45, width = 15)
for (i in 1:100){
  plot_dim1(consensus.trees100[[i]])
}
dev.off()

# now we take the majority rules consensus of the 100 resulting supertrees to only keep the nodes found in at least 50% of the trees. This should help collapse nodes that are artefacts of the random resolution of uncertain parts of the tree
con.tree.majrule <-consensus(consensus.trees100, p= 0.5, check.labels = TRUE, rooted = TRUE)
# this is a strict consensus, only containing the nodes found in every output supertree
con.tree.strict <- consensus(consensus.trees100, p=1, check.labels = TRUE, rooted = TRUE)

# for the majority rules consensus tree
plot_dim1(con.tree.majrule)
ggsave(ggtree(con.tree.majrule) + xlim(0, 45) + geom_tiplab(size=5), file = "../Output_figures/majrule_tree_100.pdf", width = 15, height=45)

# for the strict consensus tree
plot_dim1(con.tree.strict)
ggsave(ggtree(con.tree.strict) + xlim(0, 40) + geom_tiplab(size=5), file = "../Output_figures/strict_tree_100.pdf", width = 15, height=45)
```

Some differences between majority rules and strict consensus trees can be seen in the Monocalypt clade and expecially in the Radiatae series, though the general topology is the same. However, we have decided to use the strict consensus tree as the basis to which we will add species as this is the more conservative choice and it is still sufficiently resolved for our purposes.


### Renaming Known Tips

Now to edit the strict consensus tree so that the near-matched names match the trait dataset and extra subspecies can be added:

```{r fig.height=30, fig.width=15}
# starting with the tips corresponding to near-matching subspecies taxa in the traits dataset
merged.tree <- con.tree.strict
## getting rid of some erroneous node labels now that seem to interfere with the comparitive object in script 3 later
merged.tree$node.label <- NULL 

## first lets deal with the 6 aforementioned taxa from lines 342-348 and 412-419
merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus camaldulensis"] <- 'Eucalyptus camaldulensis subsp. camaldulensis'

globulus.clade <- read.tree(text = "((Eucalyptus_globulus_subsp._globulus, Eucalyptus_globulus_subsp._bicostata, Eucalyptus_globulus_subsp._pseudoglobulus, Eucalyptus_globulus_subsp._maidenii));")
merged.tree <- replace_tip(merged.tree, "Eucalyptus globulus", globulus.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus delegatensis"] <- 'Eucalyptus delegatensis subsp. delegatensis'

radiata.clade <- read.tree(text = "(Eucalyptus_radiata_subsp._radiata, Eucalyptus_radiata_subsp._robertsonii);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus radiata", radiata.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus marginata"] <- 'Eucalyptus marginata subsp. marginata'

merged.tree$tip.label[merged.tree$tip.label == "Angophora costata"] <- 'Angophora costata subsp. costata'


## now for the rest of the trait taxa with subspecies epithets but species-level matches on the thornhill tree (lines 451-468)
pauciflora.clade <- read.tree(text = "(Eucalyptus_pauciflora_subsp._pauciflora, Eucalyptus_pauciflora_subsp._acerina, Eucalyptus_pauciflora_subsp._parvifructa, Eucalyptus_pauciflora_subsp._debeuzevillei, Eucalyptus_pauciflora_subsp._niphophila, Eucalyptus_pauciflora_subsp._hedraia);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus pauciflora subsp. pauciflora", pauciflora.clade)

viminalis.clade <- read.tree(text = "(Eucalyptus_viminalis_subsp._cygnetensis, Eucalyptus_viminalis_subsp._pryoriana, Eucalyptus_viminalis_subsp._viminalis);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus viminalis subsp. viminalis", viminalis.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus calycogona subsp. calycogona"] <- 'Eucalyptus calycogona subsp. trachybasis'

cinerea.clade <- read.tree(text = "(Eucalyptus_cinerea_subsp._victoriensis, Eucalyptus_cinerea_subsp._cinerea);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus cinerea subsp. triplex", cinerea.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus tricarpa"] <- 'Eucalyptus tricarpa subsp. tricarpa'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus camphora subsp. camphora"] <- 'Eucalyptus camphora subsp. humeana'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus sideroxylon"] <- 'Eucalyptus sideroxylon subsp. sideroxylon'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus alligatrix subsp. alligatrix"] <- 'Eucalyptus alligatrix subsp. limaensis'

baueriana.clade <- read.tree(text = "(Eucalyptus_baueriana_subsp._baueriana, Eucalyptus_baueriana_subsp._thalassina);")
merged.tree <- replace_tip(merged.tree,"Eucalyptus baueriana", baueriana.clade)

leucoxylon.clade <- read.tree(text = "(Eucalyptus_leucoxylon_subsp._stephaniae, Eucalyptus_leucoxylon_subsp._pruinosa, Eucalyptus_leucoxylon_subsp._leucoxylon);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus leucoxylon subsp. leucoxylon", leucoxylon.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus polyanthemos subsp. polyanthemos"] <- 'Eucalyptus polyanthemos subsp. vestita'

tereticornis.clade <- read.tree(text = "(Eucalyptus_tereticornis_subsp._mediana, Eucalyptus_tereticornis_subsp._tereticornis);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus tereticornis subsp. tereticornis", tereticornis.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus ligulata subsp. ligulata"] <- 'Eucalyptus ligulata subsp. stirlingica'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus decipiens subsp. decipiens"] <- 'Eucalyptus decipiens subsp. chalara'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus conglobata subsp. conglobata"] <- 'Eucalyptus conglobata subsp. perata'

goniocalyx.clade <- read.tree(text = "(Eucalyptus_goniocalyx_subsp._goniocalyx, Eucalyptus_goniocalyx_subsp._viridissima);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus goniocalyx subsp. goniocalyx", goniocalyx.clade)

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus lehmannii subsp. lehmannii"] <- 'Eucalyptus lehmannii subsp. parallela'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus pluricaulis subsp. pluricaulis"] <- 'Eucalyptus pluricaulis subsp. porphyrea'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus rudis"] <- 'Eucalyptus rudis subsp. rudis'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus thamnoides"] <- 'Eucalyptus thamnoides subsp. thamnoides'

merged.tree$tip.label[merged.tree$tip.label == "Eucalyptus xanthonema subsp. xanthonema"] <- 'Eucalyptus xanthonema subsp. apposita'
```


```{r fig.height=25, fig.width=15}
#now for the near-matched species-only traits taxa with members of the same species on the thorntree (lines 489-496)
sspp.to.spp <- c("Eucalyptus willisii subsp. willisii", "Eucalyptus uncinata subsp. uncinata", "Eucalyptus socialis subsp. socialis")
## checking that all of these are recognised and match tips with no spelling mistakes, hence all 3 names should be in this list
sspp.to.spp[sspp.to.spp %in% merged.tree$tip.label] 

# substituting out everything after the subsp. word for the species in the above list
merged.tree$tip.label[merged.tree$tip.label %in% sspp.to.spp] <- sub(" subsp..+$", "", merged.tree$tip.label[merged.tree$tip.label %in% sspp.to.spp])
# getting rid of pesky underscores
merged.tree$tip.label <- gsub("_", " ", merged.tree$tip.label)
```


Now to handle those of the 20 mismatches left that we have agreed solutions for:


```{r}
# the rule-taxa with basionyms that were sister subspecies to other taxa on the tree (and hence are using the species they used to be taxonomically treated as subspecies to taxa existing on the tree)
falciformis.clade <- read.tree(text = "(Eucalyptus_falciformis, Eucalyptus_willisii);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus willisii", falciformis.clade)

sabulosa.clade <- read.tree(text = "(Eucalyptus_aromaphloia, Eucalyptus_sabulosa);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus aromaphloia", sabulosa.clade)

rossii.clade <- read.tree(text = "(Eucalyptus_rossii, Eucalyptus_racemosa);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus racemosa", rossii.clade)

merged.tree$tip.label <- gsub("_", " ", merged.tree$tip.label)

all.taxa.clean$taxon[!all.taxa.clean$taxon %in% merged.tree$tip.label] #once E. costata subsp. murrayana is dealt with (below) then there are only the 16 taxa (11 plus the 3 outgroup taxa + 2 taxa we are about to place)
```

### Which close relatives to use to place remaining taxa?

See what euclid says about the other 10 taxa that aren't outgroups or newly added taxa:

E. wimmerensis (only synonym E. viridis subsp. wimmerensis listed as vicflora synonym, very much a Rule-taxon) (Rule was a taxonomist particularly fond of splitting groups into separate species where many other would not have done so), 

E. silvestris (Rule-taxon and considered by VicFlora to be somewhat intermediate between E. microcarpa and E. wimmerensis, also considered to be a depauperate but distinguishable form of E. microcarpa),
and;
E. carolaniae (vicflora has it having a synonym of E. aff. cypellocarpa (Mornington Peninsula) so maybe its closer to E. cypellocarpa than it is to E. goniocalyx and can be positioned accordingly) are nowhere to be found on Euclid and;

E. bunyip (VicFlora has no synonyms listed, it is a Rule taxon, but E. strzeleckii and E. camphora are noted as taxa it can be distinguished from) is noted as a synonym but not considered distinct enough to be included in EUCLID.


E. punctata has closest relatives, some of which are on the thornhill tree:
    - E. canaliculata (tiny coastal range north of Sydney)
    - E. grisea (tiny range in central QLD)
    - E. propinqua (similar range though not quite as south-reaching)
    - E. major (medium range all contained within QLD; potentially subspecies of E. propinqua)

```{r}
which(thorn.tree$tip.label == "Eucalyptus canaliculata")
which(thorn.tree$tip.label == "Eucalyptus grisea")
which(thorn.tree$tip.label == "Eucalyptus propinqua")
which(thorn.tree$tip.label == "Eucalyptus major")
```

They are relatively close on the thornhill tree to begin with and so it is likely they may even become sister to each other after it is pruned.


Where was E. punctata collected? Maybe I can use which of these relatives might be collected in a similar area.

```{r}
unique(combined_traits[combined_traits$taxon == "Eucalyptus punctata", c("locality")])
```

Samples were collected at the very southern tip of this species range, just south of Sydney and not far inland. Doesn't immediately make one of these close relatives seem more representative than the others because, although it knocks out E. grisea and E. major (both restricted to QLD), it could still be either E. canaliculata or E. propinqua, which are still in slightly different places on the tree.


E. imitans a close relative of E. baxteri, which is on the traits dataset and already in the tree!

E. chapmaniana does not have any close relatives mentioned beyond how to tell it apart from certain other species.

E. behriana has close relatives E. populnea and E. sparsa (eastern and central australia repsectively). Can potentially say our sample is more likely to be closer to E. populnea based on the range in which it was collected (which was in SE Vic, Long Forest)

```{r}
# checking how close together these close relatives of E. behriana are, they appear quite far from one another. Also, it appears that tip label number is not always in the order they appear when the tree is plotted so we can also verify how far away these taxa are by looking at the pdf of the thornhill tree, which confirms it
which(thorn.tree$tip.label == "Eucalyptus populnea")
which(thorn.tree$tip.label == "Eucalyptus sparsa")

# now to look at where our amples of E. behriana were collected to see if that sheds any light on the matter
unique(combined_traits[combined_traits$taxon == 'Eucalyptus behriana', c('locality')])
```

E. strzeleckii closest to E. yarraensis, E. ovata and E. brookeriana (latter has its autonym ssp. on the thornhill tree and the other are allready on the tree)

E. arenicola was once considered part of E. willisii subsp. willisii so can probably be placed as such nested within it or, more conservatively, as a polytomy with E. falciformis as well.

Now to add the 2 (that EUCLID named the closest relative of) accordingly:

```{r}
imitans.clade <- read.tree(text = "(Eucalyptus_imitans, Eucalyptus_baxteri);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus baxteri", imitans.clade)

arenicola.clade <- read.tree(text = "(Eucalyptus_falciformis, Eucalyptus_willisii, Eucalyptus_arenicola);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus willisii", arenicola.clade)
merged.tree <- drop.tip(merged.tree, c("Eucalyptus falciformis"))
```

Now also to remember to add E. costata murrayana as a sister taxon to E. incrassata, since Nicolle treats them as a single taxon indicating they are likely closer to each other than they are to other taxa, though APC considers them to both be current names.

```{r}
murrayana.clade <- read.tree(text = "(Eucalyptus_incrassata, Eucalyptus_costata_subsp._murrayana);")
merged.tree <- replace_tip(merged.tree, "Eucalyptus incrassata", murrayana.clade)
merged.tree$tip.label <- gsub("_", " ", merged.tree$tip.label)
```


Now that we know what to keep as potential placeholders for these other 10 non-outgroup taxa (E. punctata - 1st 3, E. wimmerensis - next 1 (2nd one already on tree), E. silvestris - relative of E. wimmerensis and E. microcarpa latter on the tree allready, E. behriana - next 2, E. strzeleckii - E. brookeriana sp., E. dendromorpha - last 2 before outgroup, E. chapmaniana, E. robusta, E. carolaniae, E. bunyip - these last 4 we're not sure where to put them yet), we can prune away the unecessary bayly species that aren't in the trait dataset. Note we keep one of the outgroups on the tree to help position our outgroups on an existing node.

```{r fig.height=30, fig.width=15}
extra.taxa.for.keeping2 <- data.frame("taxon" = c("Eucalyptus canaliculata", "Eucalyptus grisea", "Eucalyptus major", "Eucalyptus polybractea", "Eucalyptus populnea", "Eucalyptus sparsa", "Eucalyptus brookeriana subsp. brookeriana", "Eucalyptus apiculata", "Eucalyptus burgessiana", "Stockwellia quadrifida"))
# just checking the list above has no spelling mistakes and they're all on the taxon list for the current growing tree we're working with
extra.taxa.for.keeping2[!extra.taxa.for.keeping2$taxon %in% merged.tree$tip.label,]

merged.tree$tip.label <- gsub("_", " ", merged.tree$tip.label)
taxa.for.keeping2 <- data.frame("taxon" = rbind(all.taxa.clean, extra.taxa.for.keeping2))
# checking there are no duplicates, there appears to be none
sum(duplicated(taxa.for.keeping2$taxon))

# now to prune away all the close relative we kept on the thornhill tree that we know we are no longer going to be using as we've narrowed some of them down. Note: taxa that had a match at species level on th thornhill tree have at this point been changed to match the version of the name that is in the trait dataset, hence why all the taxa we need are in the taxa.for.keeping2 object with the trait dataset list plus the much smaller list of close relatives.
merged.pruned.tree <- keep.tip(merged.tree, merged.tree$tip.label[merged.tree$tip.label %in% taxa.for.keeping2$taxon])

merged.pruned.tree$tip.label[merged.pruned.tree$tip.label == "Stockwellia quadrifida"] <- "Syncarpia glomulifera subsp. glomulifera"
plot_dim1(merged.pruned.tree)
```


Also, according to Thornhill, tribes Leptospermeae and Chameliauciae are closest to Eucalypteae followed by Syncarpiae. Tristaniopsis and Lophostemon are in their own different tribes so of these three aforementioned tribes that are actually closest to Eucalypts, only the Syncarpiae will have a member that we have data for and can be the closest outgroup. Hence, that can help us resolve the basal node of the tree by placing non-syncarpiae taxa as outgroups to the clade: (Syncapia + (Angophora + Corymbia + Eucalyptus)). None of the outgroups on the Thornhill or bayly trees are members of any of these other tribes so they can't be used for any positioning either.
Also, Lophostemon has been chosen to be the most preferable outgroup of the remaining two non-Syncarpia outgroup taxa due to Tristaniopsis having only a single relatively small individual measured and so we have better data for Lophostemon confertus and will use that as the outgroup instead.

```{r fig.height=30, fig.width=15}
# Now to organise the outgroups with Syncarpia as the nearest non-eucalypt taxon
## creates a tree of two taxa with new outgroup to be added and the outermost outgroup on the current tree so the function knows where to attach the extra taxon
final.tree1 <- read.tree(text = "((Lophostemon_confertus, Syncarpia_glomulifera_subsp._glomulifera));")
## sets root edge of the larger pruned merged tree to 0 so the binding works properly
merged.pruned.tree$root.edge <- 0
## binds the smaller tree as the new outermost outgroup to the tree but binds both the the desired additional taxon and the one used to guide the position of binding so now there are duplicates of the latter taxon
final.tree1 <- bind.tree(final.tree1, merged.pruned.tree, where = 'root')
## removing the duplicate taxon, which we can identify because the newly added taxa have underscores whereas the taxa from the larger pruned merged tree all had spaces a separators
final.tree1 <- drop.tip(final.tree1, c("Syncarpia_glomulifera_subsp._glomulifera"))
## now to remove the underscores
final.tree1$tip.label <- gsub("_", " ", final.tree1$tip.label)
plot_dim1(final.tree1)

# to annotate the possibilities
pdf("../Output_figures/final_tree_1_for_annotation.pdf", width = 20, height=45)
plot_dim1(final.tree1)
dev.off()
```
  
  
  
### Testing Placement Effects   

So, at this point there are E. punctata (either next to propinqua or next to the more nested one of the other group), E. behriana that we will place in various combinations to see if it changes phylogenetic signal or some correlations we know to be sensitive to phylogenetic signal.
Should we also do that with a few others such as E. dendromorpha (4 possible positions that are all different), E. wimmerensis (2 possible positions), E. silvestris (near wimmerensis or E. microcarpa), E. carolaniae (either near cypellocarpa or near goniocalyx), E. bunyip (near camphora subsp. humeana or near strzeleckii), E. strzeleckii (near one of the peppermints; ovata, yarraensis, brookeriana)

Also note which of these, if any, have extreme values for traits:
E. behriana has rather low values for reproductive traits but rather high stem density so can skew phylo-signal probably the most 
E. punctata mostly intermediately-valued (unlikely to cause great fluctuations in phylogenetic signal based on where it is positioned)
E. wimmerensis has very small leaves, but intermediate SLA and missing data for most of the remaining traits
E. silvestris has somehwhat low values for most traits but not particularly extreme values for them
E. carolaniae has quite large leaves, though intermediate everything else
E. bunyip has quite small reproductive characteristics but larger in everything else. This taxon might also be more problematic however, since one its close relatives is itself a taxon we are somewhat unsure of so it may not be worth compounding the errors.
E. strzeleckii has biggish leaf-traits and very small reproductive ones
E. chapmaniana has some large leaves in there
E. dendromorpha has relatively intermediate trait values, however there are no reproductive traits for this species recorded at all and only 2 individuals were measured for the remaining traits so we're not including this species as the extra uncertainty from inferring its position is likely greater than the gain of having an extra data point for analysis.

Hence, some combinations most useful to try:
- behriana next to largiflorens vs. next to populnea (sparsa isn't that far from populnea anyway so I can probabaly leave that one)
- wimmerensis (either next to viridis vs. next to polybractea)
- silvestris next to wimmerensis and microcarpa (which is sister to polybractea anyway) or not at all 
- punctata (either next to propinqua or next to the nested one; keep behriana variable but have the others either both included or no silvestris)

```{r fig.height=30, fig.width=15}
# for the first combination:
  #behriana-largiflorens and wimmerensis-polybractea
behriana.clade1 <- read.tree(text = "(Eucalyptus_behriana, Eucalyptus_largiflorens);")
pot.tree1 <- replace_tip(final.tree1, "Eucalyptus largiflorens", behriana.clade1)
wimm.clade1 <- read.tree(text = "(Eucalyptus_polybractea, Eucalyptus_wimmerensis);")
pot.tree1 <- replace_tip(pot.tree1, "Eucalyptus polybractea", wimm.clade1)
pot.tree1$tip.label <- gsub("_", " ", pot.tree1$tip.label)

#for behriana-largiflorens and wimmerensis-viridis
pot.tree2 <- replace_tip(final.tree1, "Eucalyptus largiflorens", behriana.clade1)
wimm.clade2 <- read.tree(text = "(Eucalyptus_wimmerensis, Eucalyptus_viridis);")
pot.tree2 <- replace_tip(pot.tree2, "Eucalyptus viridis", wimm.clade2)
pot.tree2$tip.label <- gsub("_", " ", pot.tree2$tip.label)

#for behriana-populnea and wimm-poly
behriana.clade2 <- read.tree(text = "(Eucalyptus_populnea, Eucalyptus_behriana);")
pot.tree3 <- replace_tip(final.tree1, "Eucalyptus populnea", behriana.clade2)
pot.tree3 <- replace_tip(pot.tree3, "Eucalyptus polybractea", wimm.clade1)
pot.tree3$tip.label <- gsub("_", " ", pot.tree3$tip.label)

#for behriana-populnea and wimm-viridis
pot.tree4 <- replace_tip(final.tree1, "Eucalyptus populnea", behriana.clade2)
pot.tree4 <- replace_tip(pot.tree4, "Eucalyptus viridis", wimm.clade2)
pot.tree4$tip.label <- gsub("_", " ", pot.tree4$tip.label)

#for silvestris added onto where wimmerensis is with behriana with largiflorens
silvestris.clade <- read.tree(text = "(Eucalyptus_silvestris, Eucalyptus_wimmerensis, Eucalyptus_microcarpa);")
pot.tree5 <- replace_clade(pot.tree1, 'Eucalyptus wimmerensis', 'Eucalyptus microcarpa', silvestris.clade)
pot.tree5$tip.label <- gsub("_", " ", pot.tree5$tip.label)

#for silvestris and wimmerensis in the same place but behriana with populnea
pot.tree6 <- replace_clade(pot.tree3, "Eucalyptus wimmerensis", "Eucalyptus microcarpa", silvestris.clade)
pot.tree6$tip.label <- gsub("_", " ", pot.tree6$tip.label)

#for adding in punctata next to propinqua but behriana-largiflorens
punctata.clade1 <- read.tree(text = "(Eucalyptus_propinqua, Eucalyptus_punctata);")
pot.tree7 <- replace_tip(pot.tree1, "Eucalyptus propinqua", punctata.clade1)
pot.tree7$tip.label <- gsub("_", " ", pot.tree7$tip.label)

#for adding punctata next to canaliculata but behriana-largiflorens
punctata.clade2 <- read.tree(text = "(Eucalyptus_canaliculata, Eucalyptus_punctata);")
pot.tree8 <- replace_tip(pot.tree1, 'Eucalyptus canaliculata', punctata.clade2)
pot.tree8$tip.label <- gsub("_", " ", pot.tree8$tip.label)

#for adding punctata-propinqua but behriana-populnea
pot.tree9 <- replace_tip(pot.tree3, 'Eucalyptus propinqua', punctata.clade1)
pot.tree9$tip.label <- gsub("_", " ", pot.tree9$tip.label)

#for adding punctata-canaliculata but behriana-populnea
pot.tree10 <- replace_tip(pot.tree3, "Eucalyptus canaliculata", punctata.clade2)
pot.tree10$tip.label <- gsub("_", " ", pot.tree10$tip.label)

#for tree without behriana (keep wimmerensis and punctata and silvestris)
pot.tree11 <- pot.tree5 <- replace_clade(pot.tree9, 'Eucalyptus wimmerensis', 'Eucalyptus microcarpa', silvestris.clade)
pot.tree11 <- drop.tip(pot.tree11, "Eucalyptus behriana")

#for the entire null tree
pot.null <- final.tree1

# need to add an extra 2 trees adding E. chapmaniana, E. strzeleckii, and E. carolaniae. Lets try one with all three species in thir root-most possible positions, then the other with E. strzeleckii at its tip-most position next to E. ovata and E. chapmaniana at E. glaucescens and no carolaniae at all
strzeleckii_clade1 <- read.tree(text = "(Eucalyptus_yarraensis, Eucalyptus_strzeleckii);")
strzeleckii_clade2 <- read.tree(text = "(Eucalyptus_ovata_var._ovata, Eucalyptus_strzeleckii);")
carolaniae_clade <- read.tree(text = '(Eucalyptus_goniocalyx_subsp._goniocalyx, Eucalyptus_carolaniae);')
chapmaniana_clade1 <- read.tree(text = '(Eucalyptus_smithii, Eucalyptus_chapmaniana);')
chapmaniana_clade2 <- read.tree(text = '(Eucalyptus_glaucescens,  Eucalyptus_chapmaniana);')

pot.tree12 <- replace_tip(pot.null, 'Eucalyptus yarraensis', strzeleckii_clade1)
pot.tree12 <- replace_tip(pot.tree12, 'Eucalyptus smithii', chapmaniana_clade1)
pot.tree12 <- replace_tip(pot.tree12, 'Eucalyptus goniocalyx subsp. goniocalyx', carolaniae_clade)
pot.tree12$tip.label <- gsub("_", " ", pot.tree12$tip.label)

pot.tree13 <- replace_tip(pot.null, 'Eucalyptus ovata subsp. ovata', strzeleckii_clade2)
pot.tree13 <- replace_tip(pot.tree13, 'Eucalyptus glaucescens', chapmaniana_clade2)
pot.tree13$tip.label <- gsub("_", " ", pot.tree13$tip.label)

#for behriana next to sparsa
behriana.clade3 <- read.tree(text = "(Eucalyptus_behriana, Eucalyptus_sparsa);")
## need to get rid of the previous E. behriana already on pot.tree5 (next to populnea)
pot.tree14 <- drop.tip(pot.tree5, which(pot.tree5$tip.label == 'Eucalyptus behriana'))
pot.tree14 <- replace_tip(pot.tree14, "Eucalyptus sparsa", behriana.clade3)
pot.tree14$tip.label <- gsub("_", " ", pot.tree14$tip.label)

# now to prune all trees down to only the trait taxa and not the extra taxa included for placing (the pgls joining data into comparative objects does this anyway before running the analysis but i'm doing it here anyway so I can better see the effects of repositioning and what the tree used in each analysis actually looks like)
pot.trees <- c(pot.tree1, pot.tree2, pot.tree3, pot.tree4, pot.tree5, pot.tree6, pot.tree7, pot.tree8, pot.tree9, pot.tree10, pot.tree11, pot.tree12, pot.tree13, pot.tree14, pot.null)

# dunction specifically for the instance where we now need to remove all tips that were extra taxa for keeping for positioning
drop_non_trait_taxa <- function(tree) {
  drop.tip(tree, tree$tip.label[!tree$tip.label %in% all.taxa.clean$taxon])
}

# prunes all 14 potential trees down to only the taxa we have trait data for and not any that wer ekept purely to help position some of the more uncertain taxa
pot.trees.pruned <- lapply(pot.trees, drop_non_trait_taxa)

# checking if the trees are rooted, which is a requirement of subsequent analyses, they all appear to be rooted
lapply(pot.trees.pruned, is.rooted)

# good to have a copy of them all for looking at the options
pdf("../Output_figures/Potential_trees_and_results/potential_14_trees.pdf", height = 45, width = 15)
for (i in 1:15){
  plot_dim1(pot.trees.pruned[[i]])
}
dev.off()

# Saving them in a format they can be red back into R in the script being used to generate 
pot.trees.pruned.nwk <- lapply(pot.trees.pruned, write.tree)
saveRDS(pot.trees.pruned.nwk, file = "../Input_data/Potential_trees_nwk_list.RData")
```

This list of 14 trees are not all combinations f every possible position of each of the uncertain taxa, but should encompass the greatest variation in position for taxa that have particularly extreme values for particular traits. Hence, this list can be used now to determine the phylogenetic signal for all 10 traits in the final dataset and what the value of correlations were for SLA-LA and SD-MH (these were two relationships whose statistical significance we know from preliminary analysis to have been sensitive to whether or not phylogeny was accounted for).
This will have to be run through scripts 3 and 4 to actually look at the results for each of these trees.


### Finalising Tree

So after determining: that the different positions of even the most extreme-valued dubious taxa (those whose position had been narrowed down at least to 2-4 positions) didn't greatly affect the measures of phylogenetic signal; deciding its ok to only use Patrick Fahey's tree to place taxa not already on the thornhill tree (and otherwise ignore its mild conflict about other taxa that are on it already), and deciding that it's ok to exclude E.dendromorpha and E. robusta bc they only have 2-4 individuals measured and no reproductive trait data, we need to place wimmerensis, silvestris and behriana where Patrick's tree says they go.


Also, according some further reading:

Upon looking at the Jones et al. (2016) phylogeny of the Symphyomrts, E. punctata looks way more like it should go next to E. canaliculata and E. grisea (and E. longirostrata) rather than E. propinqua or E. major, so maybe it should be placed there instead.

Jones et al. put E. strzeleckii (paraphyletic w.r.t. E. cadens, though latter is not in our dataset) next to E. yarraensis (monophyletic), but should check Rule's reference anyway

E. chapmaniana may be able to be put next to E. smithii (Jones et al., 2016), but it is also very close to E. glaucescens, which is in a different position on the tree. I think the same argument (about why we can go with Patrick's tree only for the taxa not already on our tree via more direct and reliable methods), to justify why we can ignore that that these two sources we're using have major conflicts in other parts of the tree even when those conflicts are only a taxon or two removed from the site/clade of interest that we're using that source to resolve.

E. carolaniae and E. bunyip not on the Jones et al. tree so must find Rule's reference in order to determine where to place them if not entirely leaving them out
 
Now the final list of taxa we are excluding from the treee, despite being in the trait dataset is:
- Tristaniopsis laurina
- E. dendromorpha
- E. robusta
- E. bunyip
- E. carolaniae

```{r fig.height = 25, fig.width = 15}
# best to make this next tree from the original final.tree1/pot.null tree

behriana.clade <- read.tree(text = "(Eucalyptus_behriana, Eucalyptus_largiflorens);")
final.tree2 <- replace_tip(final.tree1, "Eucalyptus largiflorens", behriana.clade)

wimm.clade <- read.tree(text = "(Eucalyptus_polybractea, Eucalyptus_wimmerensis);")
final.tree2 <- replace_tip(final.tree2, "Eucalyptus polybractea", wimm.clade)

silvestris.clade <- read.tree(text = "(Eucalyptus_silvestris, Eucalyptus_wimmerensis, Eucalyptus_microcarpa);")
final.tree2 <- replace_clade(final.tree2, 'Eucalyptus_wimmerensis', 'Eucalyptus microcarpa', silvestris.clade)

punctata.clade <- read.tree(text = "(Eucalyptus_canaliculata, Eucalyptus_punctata);")
final.tree2 <- replace_tip(final.tree2, "Eucalyptus canaliculata", punctata.clade)

strzeleckii.clade <- read.tree(text = "(Eucalyptus_strzeleckii, Eucalyptus_yarraensis);")
final.tree2 <- replace_tip(final.tree2, "Eucalyptus yarraensis", strzeleckii.clade)

chapmaniana.clade <- read.tree(text = "(Eucalyptus_chapmaniana, Eucalyptus_smithii);")
final.tree2 <- replace_tip(final.tree2, "Eucalyptus smithii", chapmaniana.clade)

final.tree2$tip.label <- gsub("_", " ", final.tree2$tip.label)

# now to have a look at the current status and check nothing has gone wrong woth the names and they're all in the correct places
plot_dim1(final.tree2)

# now to prune away the close relative we kept for positioning but aren't in the trait dataset

## should be 169 taxa in the unpruned final.tree2
length(final.tree2$tip.label) 

final.tree <- keep.tip(final.tree2, final.tree2$tip.label[final.tree2$tip.label %in% all.taxa.clean$taxon])

## now the tree should have 160 tips corresponding to the 160/165 trait taxa that we have managed to determine a reasonably position for
length(final.tree$tip.label)

# final look at final product
plot_dim1(final.tree)
```

```{r, fig.height = 25, fig.width = 15}
# to export a pdf of this final tree
ggsave(plot_dim1(final.tree), file = '../Output_figures/final_tree.pdf', width = 15, height = 25)

# to export the newick file of the tree
write.tree(final.tree, file = "../Input_data/final_tree.txt") #lets put it here for now
```

### Labelling Major Groups

```{r fig.height = 25, fig.width = 15}
# we need to know the number of the nodes we wish to label
pdf("../Output_figures/nodelabels_test.pdf", height = 45, width = 15)
plot(final.tree)
nodelabels()
dev.off()

edge <- as.data.frame(final.tree$edge)

annotated.tree <- ggtree(final.tree) + xlim(0, 32) + geom_tiplab() + 
  geom_point2(aes(subset=(node==167)), shape=23, size=6, fill='red') + geom_text(aes(0.3, 17), label = 'Tribe\n Eucalypteae', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==170)), shape=23, size=5, fill='red') + geom_text(aes(1.95, 76), label = 'Subg. Symphyomyrtus', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_text(aes(10, 63), label = 'Subg. Alveolata', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==264)), shape=23, size=5, fill='red') + geom_text(aes(12.5, 18), label = 'Subg. Eucalyptus', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_text(aes(18, 9), label = 'Subg. Eudesmia', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==280)), shape=23, size=5, fill='black') + geom_text(aes(20, 54), label = 'Ser. Radiatae', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==237)), shape=23, size=5, fill='black') + geom_text(aes(9, 98), label = 'Sect. Adnataria', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==220)), shape=23, size=5, fill='black') + geom_text(aes(12.3, 86.5), label = 'Sect. "Glandulosae"', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==211)), shape=23, size=5, fill='black') + geom_text(aes(16.7, 123), label = 'Sect. Exsertaria', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==176)), shape=23, size=5, fill='black') + geom_text(aes(8.5, 131), label = 'Sect. Maidenaria', check_overlap = TRUE, color = 'black', size = 6)

ggsave(annotated.tree, file = "../Output_figures/labelled_tree.pdf", height = 25, width = 15)
```


## Heatmaps

### Heatmap of All Data

Now to plot visualise this data as a heatmap, plotting 3 different heatmaps for each group of traits: max height and stem sapwood density; SLA, leaf area and leaf mass; and fruit wall width, fruit mass, seed mass. This is so it easier to interpret patterns and read the map because it creates spaces between blocks of 2-3 related traits. Also data was log scaled (base 10) so that comparisons between traits with very different variances could be made.

```{r fig7, fig.height = 25, fig.width = 12}
# reading in the data from script 1
eucs.scaled.data <- read.csv("../Input_data/eucs_scaled_data.csv")
eucs.scaled.data <- eucs.scaled.data[,!names(eucs.scaled.data) %in% c('X')]
eucs.scaled.data$Sprouting_Type <- as.factor(gsub("_", " ", eucs.scaled.data$Sprouting_Type))
eucs.scaled.data$Sprouting_Type2 <- as.factor(gsub("_", " ", eucs.scaled.data$Sprouting_Type2))

# to read tree back in so we don't have to rerun the whole thing to rerun this section
final.tree <- read.tree("../Input_data/final_tree.txt")
final.tree$tip.label <- gsub("_", " ", final.tree$tip.label)

#preparing the data for attaching to the tree as a heatmap (i.e. need the taxa to be row names instead of the first column)
heatmap.data.height <- data.frame("MH" = eucs.scaled.data$max_height_m, "RH" = eucs.scaled.data$relative_height_by_girth, row.names = eucs.scaled.data$taxon)

heatmap.data.stem <- data.frame("SD" = eucs.scaled.data$stem_density_g_per_ml, "RBT" = eucs.scaled.data$relative_bt_by_girth, row.names = eucs.scaled.data$taxon)

heatmap.data.leaf <- data.frame("SLA" = eucs.scaled.data$sla_mm2_per_mg, "LA" = eucs.scaled.data$leaf_area_cm2, "LM" = eucs.scaled.data$leaf_mass_g, row.names = eucs.scaled.data$taxon)

heatmap.data.fruits <- data.frame("FW" = eucs.scaled.data$fruit_wall_width_mm, "FM" = eucs.scaled.data$fruit_mass_mg, "SM" = eucs.scaled.data$seed_mass_mg, row.names = eucs.scaled.data$taxon)

# the baseline tree with enough space to the right to fit all of the heatmaps to be attached
map.tree <- ggtree(final.tree, branch.length='none', layout='rectangular') %<+% eucs.scaled.data + xlim(0,70) + geom_tiplab(size = 5, offset = 26) +  geom_tippoint(aes(colour=Sprouting_Type), size=2.7) + labs(color = "Post-Fire\nRegeneration Strategy") + 
  theme(legend.title = element_text(size = 22), legend.text = element_text(size = 18)) + geom_point2(aes(subset=(node==170)), shape=23, size=5, fill='red') + geom_text(aes(1, 77), label = 'Subg.\n Symphyomyrtus', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_text(aes(10, 63), label = 'Subg. Alveolata', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==264)), shape=23, size=5, fill='red') + geom_text(aes(12, 19), label = 'Subg.\n Eucalyptus', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_text(aes(18, 9.25), label = 'Subg. Eudesmia', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==280)), shape=23, size=5, fill='black') + geom_text(aes(20, 53), label = 'Ser.\n Radiatae', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==237)), shape=23, size=5, fill='black') + geom_text(aes(9.25, 99.5), label = 'Sect.\n Adnataria', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==220)), shape=23, size=5, fill='black') + geom_text(aes(11, 87.5), label = 'Sect.\n "Glandulosae"', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==211)), shape=23, size=5, fill='black') + geom_text(aes(16, 122.5), label = 'Sect.\n Exsertaria', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==176)), shape=23, size=5, fill='black') + geom_text(aes(6.5, 131), label = 'Sect. Maidenaria', check_overlap = TRUE, color = 'black', size = 6)


# adding first heatmap of stem traits to the tree closest to the ends of the 
map.height <- gheatmap(map.tree, heatmap.data.height,
  offset = -0.5,
  width = 0.2,
  low = "green",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Value"
) + xlim(0,70)

# see if the parameters are correct, everything visiable on the page etc.
#plot(map.height)

map.stem <- gheatmap(map.height, heatmap.data.stem,
  offset = 4.5,
  width = 0.2,
  low = "green",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Value"
) + xlim(0,70)

#plot(map.stem)

map.stem.leaf <- gheatmap(map.stem, heatmap.data.leaf,
  offset = 9.5,
  width = 0.3,
  low = "green",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Value"
) + xlim(0,70)

#plot(map.stem.leaf)

map.total <- gheatmap(map.stem.leaf, heatmap.data.fruits,
  offset = 17,
  width = 0.3,
  low = "green",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Value"
) + xlim(0,70)

#plot(map.total)

ggsave(map.total, file = "../Output_figures/heatmap.pdf", width = 20, height = 28)
```


### Colorblind Heatmap

Instead of using the default colours for the heatmap (red and green), we need to use different color combinations to be more colorblind-friendly.

```{r fig8, fig.height = 25, fig.width = 12}
# reading in the scaled data from script 1
eucs.scaled.data <- read.csv("../Input_data/eucs_scaled_data.csv")
eucs.scaled.data <- eucs.scaled.data[,!names(eucs.scaled.data) %in% c('X')]
  # need to give the missing values a detectable category to later assign them a color and shape
eucs.scaled.data$Sprouting_Type3[is.na(eucs.scaled.data$Sprouting_Type3)] <- "NA"
  # need to deal with the underscores in the sprouting type columns here
eucs.scaled.data$Sprouting_Type3 <- as.factor(gsub("_", " ", eucs.scaled.data$Sprouting_Type3))
eucs.scaled.data$Sprouting_Type4 <- as.factor(gsub("_", " ", eucs.scaled.data$Sprouting_Type4))

# to read tree back in so we don't have to rerun the whole thing to rerun this section
final.tree <- read.tree("../Input_data/final_tree.txt")
final.tree$tip.label <- gsub("_", " ", final.tree$tip.label)

#preparing the data for 
heatmap.data.height <- data.frame("MH" = eucs.scaled.data$max_height_m, "RH" = eucs.scaled.data$relative_height_by_girth, row.names = eucs.scaled.data$taxon)

heatmap.data.stem <- data.frame("SD" = eucs.scaled.data$stem_density_g_per_ml, "RBT" = eucs.scaled.data$relative_bt_by_girth, row.names = eucs.scaled.data$taxon)

heatmap.data.leaf <- data.frame("SLA" = eucs.scaled.data$sla_mm2_per_mg, "LA" = eucs.scaled.data$leaf_area_cm2, "LM" = eucs.scaled.data$leaf_mass_g, row.names = eucs.scaled.data$taxon)

heatmap.data.fruits <- data.frame("FW" = eucs.scaled.data$fruit_wall_width_mm, "FM" = eucs.scaled.data$fruit_mass_mg, "SM" = eucs.scaled.data$seed_mass_mg, row.names = eucs.scaled.data$taxon)

# building the base tree topology and labeling taxon names and color by resprouting strategies
map.tree <- ggtree(final.tree, branch.length='none', layout='rectangular') %<+% eucs.scaled.data + xlim(0,45) +
  geom_tiplab(size = 5, offset = 26) + 
  geom_tippoint(aes(colour = Sprouting_Type3, shape = Sprouting_Type3), size=3.5) + scale_color_manual("Post-Fire\nRegeneration Strategy", 
                    values = c("Combination Sprouter" = "#35B779FF",
                              "Stem Only" = "#31688EFF",   
                              "Lignotuber Only"="#440154FF",
                              "Obligate Seeder"="#F0C100FF",
                              "NA" = "black")) + 
  scale_shape_manual("Post-Fire\nRegeneration Strategy", 
                     values = c("Combination Sprouter" = 16, 
                                "Stem Only" = 18,
                                "Lignotuber Only" = 17, 
                                "Obligate Seeder" = 8, 
                                "NA" = 0)) + 
  guides(color = guide_legend(override.aes = list(size = 5))) + 
  theme(legend.title = element_text(size = 22), legend.text = element_text(size = 18)) +
  geom_point2(aes(subset=(node==170)), shape=23, size=5, fill='red') + geom_text(aes(1, 77), label = 'Subg.\n Symphyomyrtus', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_text(aes(10, 63), label = 'Subg. Alveolata', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==264)), shape=23, size=5, fill='red') + geom_text(aes(12, 19), label = 'Subg.\n Eucalyptus', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_text(aes(18, 9.25), label = 'Subg. Eudesmia', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==280)), shape=23, size=5, fill='black') + geom_text(aes(20, 53), label = 'Ser.\n Radiatae', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==237)), shape=23, size=5, fill='black') + geom_text(aes(9.25, 99.5), label = 'Sect.\n Adnataria', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==220)), shape=23, size=5, fill='black') + geom_text(aes(11, 87.5), label = 'Sect.\n "Glandulosae"', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==211)), shape=23, size=5, fill='black') + geom_text(aes(16, 122.5), label = 'Sect.\n Exsertaria', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==176)), shape=23, size=5, fill='black') + geom_text(aes(6.5, 131), label = 'Sect. Maidenaria', check_overlap = TRUE, color = 'black', size = 6)



plot(map.tree)

map.height <- gheatmap(map.tree, heatmap.data.height,
  offset = -0.5,
  width = 0.2,
  low = "#FFFF00FF",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Trait Value"
) + xlim(0,70)

#plot(map.height)

map.stem <- gheatmap(map.height, heatmap.data.stem,
  offset = 4.5,
  width = 0.2,
  low = "#FFFF00FF",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Trait Value"
) + xlim(0,70)

#plot(map.stem)

map.stem.leaf <- gheatmap(map.stem, heatmap.data.leaf,
  offset = 9.5,
  width = 0.3,
  low = "#FFFF00FF",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Trait Value"
) + xlim(0,70)

#plot(map.stem.leaf)

map.total <- gheatmap(map.stem.leaf, heatmap.data.fruits,
  offset = 17,
  width = 0.3,
  low = "#FFFF00FF",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 5,
  family = "",
  hjust = 0.5,
  legend_title = "Trait Value"
) + xlim(0,70)

plot(map.total)

ggsave(map.total, file = "../Output_figures/heatmap_colorblind.pdf", width = 20, height = 28)

# just to keep the original colorblind colors in case we need them
# scale_color_manual("Post-Fire\nRegeneration Strategy", values = c("Combination Sprouter" = "blue",
#                               "Lignotuber Only" = "magenta",
#                               "Obligate Seeder" = "brown", 
#                               "NA" = "black"), 
#                     na.value = "black")
```

### Without Taxon Labels

We agreed to remove the taxon labels in the in-text figure and leave the full one above for supplementary materials in response to Pauline's comment about the font being too small and the fact that it cannot be made above the minimum 2mm on the page without overlapping horribly.

```{r fig9, fig.height = 25, fig.width = 12}
# reading in the data from script 1
eucs.scaled.data <- read.csv("../Input_data/eucs_scaled_data.csv")
eucs.scaled.data <- eucs.scaled.data[,!names(eucs.scaled.data) %in% c('X')]
  # need to give the missing values a detectable category to later assign them a color and shape
eucs.scaled.data$Sprouting_Type3[is.na(eucs.scaled.data$Sprouting_Type3)] <- "NA"
 # need to deal with the underscores in the sprouting type columns here
eucs.scaled.data$Sprouting_Type3 <- as.factor(gsub("_", " ", eucs.scaled.data$Sprouting_Type3))
eucs.scaled.data$Sprouting_Type4 <- as.factor(gsub("_", " ", eucs.scaled.data$Sprouting_Type4))

# to read tree back in so we don't have to rerun the whole thing to rerun this section
final.tree <- read.tree("../Input_data/final_tree.txt")
final.tree$tip.label <- gsub("_", " ", final.tree$tip.label)

#preparing the data for 
heatmap.data.height <- data.frame("MH" = eucs.scaled.data$max_height_m, "RH" = eucs.scaled.data$relative_height_by_girth, row.names = eucs.scaled.data$taxon)

heatmap.data.stem <- data.frame("SD" = eucs.scaled.data$stem_density_g_per_ml, "RBT" = eucs.scaled.data$relative_bt_by_girth, row.names = eucs.scaled.data$taxon)

heatmap.data.leaf <- data.frame("SLA" = eucs.scaled.data$sla_mm2_per_mg, "LA" = eucs.scaled.data$leaf_area_cm2, "LM" = eucs.scaled.data$leaf_mass_g, row.names = eucs.scaled.data$taxon)

heatmap.data.fruits <- data.frame("FW" = eucs.scaled.data$fruit_wall_width_mm, "FM" = eucs.scaled.data$fruit_mass_mg, "SM" = eucs.scaled.data$seed_mass_mg, row.names = eucs.scaled.data$taxon)

# building tree topology, taxon names, and coloring tips by resprouting strategy
map.tree <- ggtree(final.tree, branch.length='none', layout='rectangular') %<+% eucs.scaled.data + xlim(0,45) + geom_tippoint(aes(colour = Sprouting_Type3, shape = Sprouting_Type3), size=3.5) + scale_color_manual("Post-Fire\nRegeneration Strategy", 
                    values = c("Combination Sprouter" = "#35B779FF",
                              "Stem Only" = "#31688EFF",   
                              "Lignotuber Only"="#440154FF",
                              "Obligate Seeder"="#F0C100FF",
                              "NA" = "black")) + 
  scale_shape_manual("Post-Fire\nRegeneration Strategy", 
                     values = c("Combination Sprouter" = 16, 
                                "Stem Only" = 18,
                                "Lignotuber Only" = 17, 
                                "Obligate Seeder" = 8, 
                                "NA" = 0)) + 
  guides(color = guide_legend(override.aes = list(size = 5))) + 
  theme(legend.title = element_text(size = 22), legend.text = element_text(size = 18)) +
  geom_point2(aes(subset=(node==170)), shape=23, size=5, fill='red') + geom_text(aes(1, 77), label = 'Subg.\n Symphyomyrtus', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_text(aes(10, 63), label = 'Subg. Alveolata', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==264)), shape=23, size=5, fill='red') + geom_text(aes(12, 19), label = 'Subg.\n Eucalyptus', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_text(aes(18, 9.25), label = 'Subg. Eudesmia', check_overlap = TRUE, color = 'red', size = 6) + 
  geom_point2(aes(subset=(node==280)), shape=23, size=5, fill='black') + geom_text(aes(20, 53), label = 'Ser.\n Radiatae', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==237)), shape=23, size=5, fill='black') + geom_text(aes(9.25, 99.5), label = 'Sect.\n Adnataria', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==220)), shape=23, size=5, fill='black') + geom_text(aes(11, 87.5), label = 'Sect.\n "Glandulosae"', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==211)), shape=23, size=5, fill='black') + geom_text(aes(16, 122.5), label = 'Sect.\n Exsertaria', check_overlap = TRUE, color = 'black', size = 6) + 
  geom_point2(aes(subset=(node==176)), shape=23, size=5, fill='black') + geom_text(aes(6.5, 131), label = 'Sect. Maidenaria', check_overlap = TRUE, color = 'black', size = 6)

#plot(map.tree)

map.height <- gheatmap(map.tree, heatmap.data.height,
  offset = -0.5,
  width = 0.14,
  low = "#FFFF00FF",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 6,
  family = "",
  hjust = 0.5,
  legend_title = "Trait Value"
) + xlim(0,45)

#plot(map.height)

map.stem <- gheatmap(map.height, heatmap.data.stem,
  offset = 3,
  width = 0.14,
  low = "#FFFF00FF",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 6,
  family = "",
  hjust = 0.5,
  legend_title = "Trait Value"
) + xlim(0,45)

#plot(map.stem)

map.stem.leaf <- gheatmap(map.stem, heatmap.data.leaf,
  offset = 6.5,
  width = 0.21,
  low = "#FFFF00FF",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 6,
  family = "",
  hjust = 0.5,
  legend_title = "Trait Value"
) + xlim(0,45)

#plot(map.stem.leaf)

map.total <- gheatmap(map.stem.leaf, heatmap.data.fruits,
  offset = 11.7,
  width = 0.21,
  low = "#FFFF00FF",
  high = "red",
  color = "white",
  colnames = TRUE,
  colnames_position = "bottom",
  colnames_angle = 0,
  colnames_level = NULL,
  colnames_offset_x = 0,
  colnames_offset_y = 0,
  font.size = 6,
  family = "",
  hjust = 0.5,
  legend_title = "Trait Value"
) + xlim(0,42)

#plot(map.total)

ggsave(map.total, file = "../Output_figures/heatmap_colorblind_abridged.pdf", width = 20, height = 28)
```
